---
title: ""
output: html_document
# params:
#   samplename: "WT-NOPFA"
#   peak_in: "/Users/homanpj/OneDrive - National Institutes of Health/Loaner/Wolin/CLIP/6-22-21-HaCaT_fCLIP/13_annotation/02_peaks/WT-NOPFA_peakannotation_complete.txt"
#   output_table: "/Users/homanpj/OneDrive - National Institutes of Health/Loaner/Wolin/CLIP/6-22-21-HaCaT_fCLIP/13_annotation/02_peaks/WT-NOPFA_peakannotation_final.txt"
#   readdepth: 5
#   PeakIdnt: "ALL" #MultiMap, Unique, all
params:
  samplename: "WT-NOPFA"
  peak_in: "/Users/homanpj/OneDrive - National Institutes of Health/Loaner/Wolin/CLIP/6-22-21-HaCaT_fCLIP/13_annotation_test/13_annotation/02_peaks/WT-NOPFA_peakannotation_complete.txt"
  output_table: "/Users/homanpj/OneDrive - National Institutes of Health/Loaner/Wolin/CLIP/6-22-21-HaCaT_fCLIP/13_annotation_test/13_annotation/02_peaks/WT-NOPFA_peakannotation_final.txt"
  readdepth: 5
  PeakIdnt: "ALL" #MultiMap, Unique, all
# params:
#   samplename: "untrans"
#   peak_in: "/Users/homanpj/OneDrive - National Institutes of Health/Loaner/Wolin/CLIP/DPR_iCLIP/080818_DPR_CLIP/14_peaks/02_peaks/untrans_peakannotation_complete.txt"
#   output_table: "/Users/homanpj/OneDrive - National Institutes of Health/Loaner/Wolin/CLIP/DPR_iCLIP/080818_DPR_CLIP/14_peaks/02_peaks/untrans_peakannotation_final.txt"
#   readdepth: 5
#   PeakIdnt: "ALL" #MultiMap, Unique, all
---
 
 <!-- # Load libraries, set theme -->
```{r envirment, include=F ,echo=F,warning=F,message=FALSE}
rm(list=setdiff(ls(), "params"))
library(rmarkdown)
library(stringr)
library(rtracklayer)
library(VariantAnnotation,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(rmarkdown)
library(ggplot2,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library("viridis",quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(edgeR,quietly = T,verbose = F)
library('GenomicFeatures',quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library('rtracklayer',quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(matrixStats,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(plyr,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(tidyr,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(fitdistrplus,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(stringr,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(data.table)
library(reshape)
library(knitr)
library(stringi)
library(BSgenome)
library(biomaRt)
library(plotly)
library(tidyr)
library(GenomicRanges)
library(RColorBrewer)
library('gplots')
library(ggpubr)
library(circlize)
library('regioneR')
library(scales)

blank_theme <- theme_minimal()+
  theme(
  axis.title.x = element_blank(),
  axis.title.y = element_blank(),
  panel.border = element_blank(),
  panel.grid=element_blank(),
  axis.ticks = element_blank(),
  plot.title=element_text(size=14, face="bold")
  )
```

<!-- # Set params, output   -->

```{r echo=F, message=FALSE, warning=FALSE, eval=T, include=F,results=F,}
peak_in=params$peak_in
samplename=params$samplename
readdepth=as.numeric(params$readdepth)
output_table=params$output_table
PeakIdnt=params$PeakIdnt
incd_rRNA=T
```

```{r echo=F, message=FALSE, warning=FALSE, eval=T, include=F,results=F,}
# Identification of CLIP peaks: `r toString(params$samplename)`    
Peaksdata2_anno = read.table(peak_in)
# unique(Peaksdata2_anno$Same_Comb_type_exon)
Peaksdata2_anno$Same_Comb_type_exon = 
    factor(Peaksdata2_anno$Same_Comb_type_exon, 
           levels = c("ncRNA", "protein_coding: exon", "Repeat Element","pseudogene",
                      "Antisense Feature","protein_coding: Intron","lncRNA","no Feature"))  
# unique(Peaksdata2_anno$Oppo_Comb_type_exon)
Peaksdata2_anno$Oppo_Comb_type_exon = 
    factor(Peaksdata2_anno$Oppo_Comb_type_exon, 
           levels = c("ncRNA", "protein_coding: exon", "Repeat Element","pseudogene",
                      "Antisense Feature","protein_coding: Intron","lncRNA","no Feature"))  
# unique(Peaksdata2_anno$Comb_type_exon_Oppo)
Peaksdata2_anno$Comb_type_exon_Oppo = 
    factor(Peaksdata2_anno$Comb_type_exon_Oppo, 
           levels = c("ncRNA", "protein_coding: exon", "Repeat Element","pseudogene",
                      "Antisense Feature","protein_coding: Intron","lncRNA","no Feature"))  
###### remove chrM clip peaks
Peaksdata2_anno=Peaksdata2_anno[Peaksdata2_anno$chr%in%'chrM'==F,] 
# Peaksdata2_anno=Peaksdata2_anno[1:10,]

```

# CLIP peak annotations: `r toString(samplename)`  
  
## Processing data
**1. Align reads Novoalign**

**2. UMI-tools remove duplicates**

**3. Select CLIP peaks**

  Identify peaks using bedtools merge on All reads (unique and Multimapping) and count number of reads that overlap with the identified peaks.
  
  Peaks with > `r toString(readdepth)` (unique + Multi-Mapped:fraction) reads were assessed.  


```{r ,fig.align='center',echo=F,warning=F,eval=T,include=T}
##todo - make function with replacement title
if (params$PeakIdnt=='UNIQUE'){
  x = ggplot(Peaksdata2_anno, aes(x=Counts_Unique)) +
    geom_histogram(bins = 30) + 
    scale_x_continuous(trans = "log2") + 
    theme_classic()+ylab("# of peaks") + 
    xlab("Peak Counts (unique)") + 
    ggtitle("Distribution of peak counts")
} else if (params$PeakIdnt=='ALL'){
  x = ggplot(Peaksdata2_anno, aes(x=Counts_fracMM)) +
    geom_histogram(bins = 30) +
    scale_x_continuous(trans = "log2") + 
    theme_classic()+ylab("# of peaks") + 
    xlab("Peak Counts (Unique + FracMM)") + 
    ggtitle("Distribution of peak counts")
}
ggplotly(x,tooltip = c("x", "y"))
```


<br/>    
<br/>    


## Identify CLIP peak Gene Location  

### Identify if CLIP peak overlaps with Intron or Exonic region   

<br>  

<!-- Intron coordinates were calculated from GTF file. -->

<!-- A second column was added to idenify if the peak also overlapped with the 5'UTR 3'UTR or CDS (Column: Featrue 2) -->

<br> 

Peaks were annotated with overlapping gene/transcript  

Peaks were annotated by whether they overlap with Host gene intron/exon region  

<br>   

### Identify CLIP peak in ncRNA   


```{r echo=F,results='asis',eval=F}
Classification_ncRNA=read.delim(paste0(OUT,"ncRNA_Annotations.txt"), header=T, sep="\t",stringsAsFactors = F)
kable(Classification_ncRNA[,1:3])
```

<br>  


### Identify peaks in repeat regions   

   Annotate all repeat regions/Classes identified in Repeatmasker Annotation file (UCSC Table browser)  
   Data was not filtered based on any of the identified Repeats.  
    1) LINE/SINE   
    2) LTR   
    3) DNA   
    4) Satellites   
    5) Simple Repeats   
    6) Low Complexity   
    7) Other   
    8) Unknown   
    <br>   


```{r echo=F,warning=F, include=T,eval=T,fig.align='center'}
p=as.data.frame(((colSums(is.na(Peaksdata2_anno[,grep('Same_Repeat_',colnames(Peaksdata2_anno))])==F)/nrow(Peaksdata2_anno))*100))
p$Repeat_Type=rownames(p);colnames(p)=c('Frequency','Repeat_Type')
p=p[rownames(p)%in%'Same_Repeat_comb'==F,]
p$Repeat_Type=gsub("Same_Repeat_","",p$Repeat_Type)
ggplot(p) +
  geom_bar(aes(x=Repeat_Type,y=Frequency), stat="identity")+theme_classic() +
  xlab("")+ylab("Frequency of peaks in \nrepeat regions (%)")+ggtitle("Same Strand") +
  theme(plot.title = element_text(hjust = 0.5,size = 15),
    axis.title=element_text(size=15),
    axis.text=element_text(size=10),
    legend.text = element_text( size=10)) +
  theme(axis.text.x = element_text(angle = -45))
```
   
<br>   
<br>   

### Asigning Clip peak attributes   

Not all Peaks overlap with a single feature so peak assignments were assigned by priority:  

<!-- **ncRNA > Protein coding : Exonic > repeats > Pseudogene > Antisense Feature > Protein Coding : Intronic > lncRNA > no Feature**    -->

**ncRNA > protein_coding: Exonic > Repeat Element > pseudogene > lncRNA: Exonic > Antisense Feature > protein_coding: Intronic > lncRNA: Intronic > no Feature**

All annotations from RNA type, Repeat regions, and Intronic/exonic regions are annotated in the Table.   

<br>  
<br>  


## Overview of targets by CLIP peak catagory   

```{r , echo=F, include=T, eval=T, fig.align='default', fig.ncol=2, fig.height=3.5, fig.width=7, out.height=200, out.width="50%", fig.show="hold", layout='l-screen', results = FALSE, message=FALSE , warning=F}
nc_color=c("ncRNA"="#1B9E77","rRNA" ="#1fb487",
           "yRNA"= "#2bdba6" ,
           "protein_coding: exon"="#D95F02", "Repeat Element"="#7570B3" ,"pseudogene"="#E7298A", "Antisense Feature"="#66A61E", "protein_coding: Intron"="#E6AB02", "lncRNA"="#A6761D", "no Feature"="#666666")
#Plot 1
if (params$PeakIdnt=='UNIQUE'){
  x1=ggplot(Peaksdata2_anno,aes(x=Counts_Unique,y=Length,color=Comb_type_exon_Oppo,text=Same_external_gene_name))+geom_point(size=1)+
    # scale_x_continuous(trans = "log2",limits = c(.5,NA))+
    scale_y_continuous(trans = "log2",limits = c(NA,NA), oob = squish_infinite) +
    theme_classic()+
    ylab("Peak Width")+xlab("Peak Counts (unique)")+ggtitle("Comparison of peak counts and Peak Width")+
 scale_color_manual(values=nc_color) 
} else if (params$PeakIdnt=='ALL'){
  x1=ggplot(Peaksdata2_anno,aes(x=Counts_fracMM,y=Length,color=Comb_type_exon_Oppo,text=Same_external_gene_name))+geom_point(size=1)+
      # scale_x_continuous(trans = "log2",limits = c(4,NA))+
      scale_y_continuous(trans = "log2",limits = c(NA,NA), oob = squish_infinite) + 
    theme_classic()+
    ylab("Peak Width")+xlab("Peak Counts (Unique+ FracMM)")+ggtitle("Comparison of peak counts and Peak Width")+  labs(color='RNA Type')+
 scale_color_manual(values=nc_color) 
}
#plot 2
x2=ggplot(Peaksdata2_anno,aes(x=Length))+geom_histogram(bins = 100)+
  # scale_x_continuous(trans = "log2",limits = c(20,NA)) + 
  theme_classic()+ylab("# of peaks")+xlab("Peak Width")+ggtitle("Distribution of peak Width")
#plot 3
x3=ggplot(Peaksdata2_anno,aes(x=Length,line=Comb_type_exon_Oppo,color=Comb_type_exon_Oppo))+geom_density(size=.5)+
  # scale_x_continuous(trans = "log2",limits = c(8,NA)) + 
  theme_classic()+ylab("Density")+xlab("Peak Width")+ggtitle("Distribution of peak Width")+  labs(color='RNA Type')+
 scale_color_manual(values=nc_color) 
#plot 4
x4=ggplot(Peaksdata2_anno,aes(x=Length,line=Comb_type_exon_Oppo,color=Comb_type_exon_Oppo))+geom_freqpoly(bins = 100)+
  # scale_x_continuous(trans = "log2",limits = c(8,NA)) + 
  theme_classic()+ylab("# of peaks")+xlab("Peak Width")+ggtitle("Distribution of peak Width")+  labs(color='RNA Type')+
 scale_color_manual(values=nc_color) 
#plot 5
if (params$PeakIdnt=='UNIQUE'){
x5=ggplot(Peaksdata2_anno,aes(x=Counts_Unique,line=Comb_type_exon_Oppo,color=Comb_type_exon_Oppo))+geom_density()+
  # scale_x_continuous(trans = "log2",limits = c(.25,NA)) + 
  theme_classic()+ylab("Density")+xlab("Peak Counts (unique)")+ggtitle("Distribution of peak counts")+ 
  labs(color='RNA Type')+
 scale_color_manual(values=nc_color) 
} else if (params$PeakIdnt=='ALL'){
  x5=ggplot(Peaksdata2_anno,aes(x=Counts_fracMM,line=Comb_type_exon_Oppo,color=Comb_type_exon_Oppo))+geom_density() +
    # scale_x_continuous(trans = "log2",limits = c(.25,NA)) + 
    theme_classic()+ylab("Density")+xlab("Peak Counts (Unique+ FracMM)")+ggtitle("Distribution of peak counts") +
    labs(color='RNA Type') +
 scale_color_manual(values=nc_color) 
}
#plot all graphs
if (nrow(Peaksdata2_anno)>1) {
  x1=x1+scale_x_continuous(trans = "log2",limits = c(4,NA));  
  x2=x2+scale_x_continuous(trans = "log2",limits = c(20,NA));  
  x3=x3+scale_x_continuous(trans = "log2",limits = c(8,NA));
  x4=x4+scale_x_continuous(trans = "log2",limits = c(8,NA));  
  x5=x5+scale_x_continuous(trans = "log2",limits = c(.25,NA))
}
  x1;x2;x3;x4;x5
```

<br>  
<br>   

##  Number of peaks assigned to each catagory  


**Peak counts do include rRNA peaks**   

```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,eval=T,include=T}
p=Peaksdata2_anno
p$PlotColumn=p$Comb_type_exon_Oppo
if (incd_rRNA==FALSE) {
  p=p[!p$Same_Comb_type_ncRNA%in%'rRNA',]
}
if (nrow(p)>0) {
#change types
p[is.na(p$PlotColumn),'PlotColumn']='no Feature'
colnames(p)[colnames(p)%in%'PlotColumn']='RNA_Type'
p$RNA=''
#plot
gg=ggplot(p,aes(fill=RNA_Type,x=RNA) ) +
  geom_bar(position="stack",width=.5) +
  theme_classic()+ggtitle("BioType: # of peaks per catagory") +
  scale_fill_manual(values=nc_color) 
ggplotly(gg)
}else{print("No ncRNA")}
```


```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,eval=T,include=T}
p=Peaksdata2_anno
p$PlotColumn=p$Comb_type_exon_Oppo
###phil need else statement
#check if this makes sense
if (incd_rRNA==FALSE) {
  p=p[!p$Same_Comb_type_ncRNA%in%'rRNA',]
} else{
  p=p
}
#filter
p=p[p$PlotColumn%in%'ncRNA','Same_Comb_type_ncRNA',drop=F]
#if there are remaining types
if (nrow(p)>0) {
  p$RNA=""
  gg=ggplot(p,aes(fill=Same_Comb_type_ncRNA,x=RNA) ) +
    geom_bar(width=.5)+theme_classic()+
    ggtitle("ncRNA only: # of peaks per catagory") +
    labs(fill='RNA Type') +
    scale_color_brewer(palette = "Paired")+scale_fill_brewer(palette = "Paired") + 
    blank_theme + 
    coord_polar("y",start=0)
  (gg) ###updating print to be consistent
  # ggplotly(gg)
}else{print("No ncRNA")}
 
```    

<br>  
<br> 

**Peak counts do not include rRNA peaks**   


 
```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,eval=T,include=T}
p=Peaksdata2_anno
p$PlotColumn=p$Comb_type_exon_Oppo
p=p[!p$Same_Comb_type_ncRNA%in%'rRNA',]
if (nrow(p)>0) {
p[is.na(p$Comb_type_exon),'PlotColumn']='no Feature'
colnames(p)[colnames(p)%in%'PlotColumn']='RNA_Type'
p$RNA=''
#plot
gg=ggplot(p,aes(fill=RNA_Type,x=RNA) )+
  geom_bar(position="stack",width=.5) +
  theme_classic() +
  ggtitle("BioType: # of peaks per catagory") +
  scale_fill_manual(values=nc_color) 
ggplotly(gg)
}else{print("No ncRNA")}
```

```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,eval=T,include=T}
p=Peaksdata2_anno
p$PlotColumn=p$Comb_type_exon_Oppo
p=p[!p$Same_Comb_type_ncRNA%in%'rRNA',]
p=p[p$PlotColumn%in%'ncRNA','Same_Comb_type_ncRNA',drop=F]
#if there are remaining types
if (nrow(p)>0) {
  p$RNA=""
  gg=ggplot(p,aes(fill=Same_Comb_type_ncRNA,x=RNA) ) +
    geom_bar(width=.5)+theme_classic()+
    ggtitle("ncRNA only: # of peaks per catagory") +
    labs(fill='ncRNA Type')+
    scale_color_brewer(palette = "Paired")+ 
    scale_fill_brewer(palette = "Paired") +  blank_theme+ 
    coord_polar("y",start=0)
  (gg)
  # ggplotly(gg)
}else{print("No ncRNA")}
```       

<br>  



## Number of reads assigned to each catagory   

#### Counts for RNA biotype  

<br>   

```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,include=F,eval=T}
p=Peaksdata2_anno
p$PlotColumn=p$Comb_type_exon_Oppo
p$PlotColumn=as.character(p$PlotColumn)
p[p$Same_Comb_type_ncRNA%in%'rRNA','PlotColumn']='rRNA'
p[p$Same_Comb_type_ncRNA%in%'yRNA','PlotColumn']='yRNA'
p$PlotColumn=factor(p$PlotColumn, levels = c("ncRNA","rRNA","yRNA", "protein_coding: exon", "Repeat Element","pseudogene","Antisense Feature","protein_coding: Intron","lncRNA","no Feature"))
u=c("ncRNA","rRNA","yRNA", "protein_coding: exon", "Repeat Element","pseudogene","Antisense Feature","protein_coding: Intron","lncRNA","no Feature")
if (nrow(p)>0) {
fcount=data.frame(u,1:length(u));colnames(fcount)=c('RNA_Type','order');fcount$Counts_Unique=NA;fcount$Counts_MM=NA;fcount$Peak_count=NA
for (x in 1:length(u)) {
  pam=p[p$PlotColumn%in%u[x],]
  fcount[x,1]=u[x]
  fcount[x,'Counts_Unique']=colSums(pam[,'Counts_Unique',drop=F],na.rm = T)
  fcount[x,'Counts_fracMM']=colSums(pam[,'Counts_fracMM',drop=F],na.rm = T)
    fcount[x,'Peak_count']=nrow(pam)
fcount$RNA=""
fcount$RNA_Type=factor(fcount$RNA_Type,
  levels = c("ncRNA","rRNA","yRNA",
              "protein_coding: exon", "Repeat Element",
              "pseudogene","Antisense Feature","protein_coding: Intron","lncRNA","no Feature"))
}
} else {print('No Features')}
```
```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,include=T,eval=T}
if (exists('fcount')) {
  
gg1 = ggplot(fcount,aes(fill=RNA_Type,x=RNA,y=Counts_Unique)) +
  geom_bar(stat='identity',position="stack",width=.5)+theme_classic() +
  ggtitle("BioType - # of reads (unique) by catagory)") +
  scale_fill_manual(values=nc_color)  
ggplotly(gg1)
} else {print('No Features')}
```

```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,include=T,eval=T}
if (exists('fcount')) {
gg2 = ggplot(fcount,aes(fill=RNA_Type,x=RNA,y=Counts_fracMM)) +
  geom_bar(stat='identity',position="stack",width=.5) +
  theme_classic() +
  ggtitle("BioType - # of reads (unique + Frac. MM) by catagory)")+
  scale_fill_manual(values=nc_color)
ggplotly(gg2)
} else {print('No Features')}
```
<br>    

#### Counts for ncRNA      


```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,include=F,eval=T}
p=Peaksdata2_anno
p$PlotColumn=p$Comb_type_exon_Oppo
p$PlotColumn=as.character(p$PlotColumn)
p[p$Same_Comb_type_ncRNA%in%'rRNA','PlotColumn']='rRNA'
p[p$Same_Comb_type_ncRNA%in%'yRNA','PlotColumn']='yRNA'
p$PlotColumn=factor(p$PlotColumn, levels = c("ncRNA","rRNA","yRNA", "protein_coding: exon", "Repeat Element","pseudogene","Antisense Feature","protein_coding: Intron","lncRNA","no Feature"))
p=p[p$PlotColumn%in%'ncRNA',]
if (nrow(p)>0) {
  u=unique(p$Same_Comb_type_ncRNA)
  nccount=data.frame(u,1:length(u));colnames(nccount)=c('Same_RNA_Subtype','Counts_All');nccount$Counts_Unique=NA;nccount$Counts_MM=NA;nccount$Peak_count=NA
  
  for (x in 1:length(u)) {
    pam=p[p$Same_Comb_type_ncRNA%in%u[x],]
    nccount[x,1]=u[x]
    nccount[x,'Counts_Unique']=colSums(pam[,'Counts_Unique',drop=F],na.rm = T)
    nccount[x,'Counts_fracMM']=colSums(pam[,'Counts_fracMM',drop=F],na.rm = T)
    nccount[x,'Peak_count']=nrow(pam)
  }
}else{print('No Significant ncRNA CLIP peaks')}
```
```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,include=T,eval=T}
if (exists('nccount')) {
nccount$RNA=''
#plot
gg1 = ggplot(nccount,aes(fill=Same_RNA_Subtype,x=RNA,y=Counts_Unique) ) +
  geom_bar(stat='identity',width=.5)+theme_classic() +
  ggtitle("ncRNA only: # of reads (unique) by catagory") + 
  labs(color='RNA Type') +
  scale_color_brewer(palette = "Paired")+scale_fill_brewer(palette = "Paired") + 
  blank_theme + 
  coord_polar("y",start=0)
plot(gg1)
}else{print('No Unique counts for ncRNA peaks')}
#plot
```
```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,include=T,eval=T}
  
if (exists('nccount')) {
nccount$RNA=''
gg2 = ggplot(nccount,aes(fill=Same_RNA_Subtype,x=RNA,y=Counts_fracMM) ) +
  geom_bar(stat='identity',width=.5)+theme_classic() +
  ggtitle("ncRNA only: # of reads (unique + Frac. MM) by catagory") + 
  labs(color='RNA Type') +
  scale_color_brewer(palette = "Paired")+scale_fill_brewer(palette = "Paired") +
  blank_theme + 
  coord_polar("y",start=0)
plot(gg2)
}else{print('No unique + Frac. MM counts for ncRNA CLIP peaks')}
```      
<br>   

<br>   

### Read Counts by peak for each catagory    
  
#### Unique Read counts (Count Unique reads only)   

<br>   

```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=8,echo=F,warning=F,eval=T, include=T}
p=Peaksdata2_anno
p$PlotColumn=p$Comb_type_exon_Oppo
if (incd_rRNA==FALSE) {
  p=p[!p$Same_Comb_type_ncRNA%in%'rRNA',]
}
colnames(p)[colnames(p)%in%'PlotColumn']='RNA_Type'
if (nrow(p)>0) {
p$RNA=''
#plot
gg = ggplot(p,aes(x=RNA_Type,y=Counts_Unique,fill=RNA_Type)) +
  geom_violin()+theme_classic()+geom_jitter(height = 0, width = 0.1,size=0.5,aes(text=ID))+
  scale_y_continuous(trans = "log2", oob = squish_infinite) +
  ggtitle('Peak Counts (Unique) for\nby Feature') +
  ylab("Counts")+xlab("") +
  theme(plot.title = element_text(hjust = 0.5,size = 15),axis.title=element_text(size=15),axis.text=element_text(size=12))+theme(legend.position = "none") +
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = -45))+
  scale_fill_manual(values=nc_color) 

if (nrow(p)==1|nrow(p)>5000) {gg
  }else { ggplotly(gg)}
}else{print('No Features')}
```  
<br/>    

<br/>    

```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=8,echo=F,warning=F,eval=T, include=T}
p=Peaksdata2_anno
p$PlotColumn=p$Same_Comb_type_ncRNA
  p=p[p$Comb_type_exon_Oppo%in%'ncRNA',]
colnames(p)[colnames(p)%in%'PlotColumn']='RNA_Type'

if (nrow(p)>0) {
  p$ID=paste0(p$ID,"\n",p$Same_gene_name_comb)
#plot
gg = ggplot(p,aes(x=RNA_Type,y=Counts_Unique,fill=RNA_Type)) +
  geom_violin()+theme_classic()+
geom_jitter(height = 0, width = 0.1,size=0.5,aes(text=ID))+
  scale_y_continuous(trans = "log2", oob = squish_infinite) +
  ggtitle('Peak Counts (Unique) for\nby ncRNA Type') +
  ylab("Counts")+xlab("") +
  theme(plot.title = element_text(hjust = 0.5,size = 15),axis.title=element_text(size=15),axis.text=element_text(size=12))+theme(legend.position = "none") +
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = -45))+
scale_fill_brewer(palette = "Paired")

if (nrow(p)==1|nrow(p)>5000) {gg
  }else { ggplotly(gg,tooltip = c("text", "y"))}
}else{print('No Features')}


```



<br>   
<br>   

#### Fractional Multimap Read counts \n (Count Unique reads + (MultiMapped read / # of different mappings) )     

Does include rRNA
<br>   

```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=8,echo=F,warning=F,eval=T, include=T}
p=Peaksdata2_anno
p$PlotColumn=p$Comb_type_exon_Oppo
colnames(p)[colnames(p)%in%'PlotColumn']='RNA_Type'
if (nrow(p)>0) {
p$RNA=''
#plot
gg=ggplot(p,aes(x=RNA_Type,y=Counts_fracMM,fill=RNA_Type)) +
    geom_violin()+theme_classic()+geom_jitter(height = 0, width = 0.1,size=0.5,aes(text=ID))+
  scale_y_continuous(trans = "log2", oob = squish_infinite)+
  ggtitle('Peak Counts (Unique + Frac. MM)\nby Feature') +
  ylab("Counts")+xlab("")+
  theme(plot.title = element_text(hjust = 0.5,size = 15),axis.title=element_text(size=15),axis.text=element_text(size=12))+theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = -45))+
  scale_fill_manual(values=nc_color) 
if (nrow(p)==1|nrow(p)>5000) {gg
  }else { ggplotly(gg)}
}else {print("No Features")}
```  
<br/>    

<br/>    

```{r fig.asp=.5,fig.align='center',fig.height=6, fig.width=8,echo=F,warning=F,eval=T}
p=Peaksdata2_anno
p$PlotColumn=p$Same_Comb_type_ncRNA
  p=p[p$Comb_type_exon_Oppo%in%'ncRNA',]
colnames(p)[colnames(p)%in%'PlotColumn']='RNA_Type'

if (nrow(p)>0) {
p$ID=paste0(p$ID,"\n",p$Same_gene_name_comb)
#plot
  gg=ggplot(p,aes(x=RNA_Type,y=Counts_fracMM,fill=RNA_Type) ) +
    geom_violin()+theme_classic()+geom_jitter(height = 0, width = 0.05,size=.5,aes(text=ID)) +
    scale_y_continuous(trans = "log2", oob = squish_infinite) +
    ggtitle('Peak Counts (Unique + Frac. MM)\nby ncRNA Type')+ylab("Counts") +
    xlab("") +
    theme(plot.title = element_text(hjust = 0.5,size = 15),axis.title=element_text(size=15),axis.text=element_text(size=12)) +
    theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle = -45))+scale_fill_brewer(palette = "Paired")#+ylim(min(p$Counts_Unique),max(p$Counts_Unique))
  if (nrow(p)==1|nrow(p)>5000) {gg
  }else { ggplotly(gg,tooltip = c("text", "y"))}
 
}else {print("No Features")}
```  
<br>   
<br>   

<br>  


## Opposite strand CLIP peak attributes   


```{r , fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F}
p=Peaksdata2_anno
p=p[p$Comb_type_exon_Oppo%in%'Antisense Feature',]
###phil need else statement
if (incd_rRNA==FALSE) {
  p=p[!p$Same_Comb_type_ncRNA%in%'rRNA',]
}
if (nrow(p)>0) {
colnames(p)[colnames(p)%in%'Oppo_Comb_type_exon']='RNA_Type'
p$RNA=''
#plot
ggplot(p,aes(fill=RNA_Type,x=RNA_Type) )+geom_bar(stat="count")+theme_classic()+
  ggtitle("Opposite Strand Feature - Number of peaks by catagory") +
  xlab("")+
  theme(plot.title = element_text(hjust = 0.5),axis.title=element_text(size=15),axis.text=element_text(size=10),legend.position = "none",axis.text.x = element_text(angle = -45,vjust = .5))+
  scale_fill_manual(values=nc_color) 
}else {print("No Features")}
```

```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T,incude=T}
p=Peaksdata2_anno
##phil need else statement
if (incd_rRNA==FALSE) {
  p=p[!p$Same_Comb_type_ncRNA%in%'rRNA',]
}
p=p[p$Comb_type_exon_Oppo%in%'Antisense Feature',]
if (nrow(p)>0) {
  p$transition=paste0(p$Same_Comb_type_exon,'->\n',p$Oppo_Comb_type_exon)
colnames(p)[colnames(p)%in%'transition']='RNA_Type'
p$Difference=''
#plot
gg = ggplot(p,aes(fill=Same_Comb_type_exon,x=factor(Oppo_Comb_type_exon)) ) +
  geom_bar(stat="count")+theme_classic()+
  ggtitle("Opposite Strand Feature Features: \n Compare Antisense and Snese stand Annotation")+xlab("Antisense strand Annotation")+ylab("Peak Count")+labs(fill = "Sense Stand Annotation")+
  theme(plot.title = element_text(hjust = 0.5,size=20),axis.title=element_text(size=15),axis.text=element_text(size=10),legend.position = "right",axis.text.x = element_text(angle = -45,vjust = .5))
ggplotly(gg)
}else {print("No Features")}
```

<br>  

```{r echo=F,warning=F, include=T,eval=T}
p=as.data.frame(((colSums(is.na(Peaksdata2_anno[,grep('Oppo_Repeat_',colnames(Peaksdata2_anno))])==F))))
p$Repeat_Type=rownames(p);colnames(p)=c('Frequency','Repeat_Type')
p=p[rownames(p)%in%'Oppo_Repeat_comb'==F,]
if (nrow(p)>0) {
p$Repeat_Type=gsub("Oppo_Repeat_","",p$Repeat_Type)
gg = ggplot(p) +
  geom_bar(aes(x=Repeat_Type,y=Frequency),stat="identity") +
  theme_classic()+#ggtitle("SNP Mutations")+
  xlab("")+ylab("number of peaks in \nrepeat regions")+ggtitle("Opposite Strand")+
  theme(plot.title = element_text(hjust = 0.5,size = 15),
    axis.title=element_text(size=15),axis.text=element_text(size=10),
    legend.text = element_text( size=10)) +
  theme(axis.text.x = element_text(angle = -45))
  if (nrow(p)==1|nrow(p)>5000) {gg
  }else { ggplotly(gg)}
}else {print("No Features")}
```

```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T,include=F}
#RUN2
if ('IDmerge'%in% colnames(Peaksdata2_anno)) {
  Peaksdata3_anno=Peaksdata2_anno[,c("ID","IDmerge",'chr',"start","end","strand",
                                'Counts_Unique','Counts_fracMM',
                                "Same_gene_name_comb","Same_ensembl_gene_id","Same_type_comb","Same_Repeat_comb",'Same_feature',"Same_Comb_type_exon","Same_Comb_type_ncRNA","Oppo_gene_name_comb","Oppo_ensembl_gene_id","Oppo_type_comb","Oppo_Repeat_comb",'Oppo_feature',"Oppo_Comb_type_exon","Oppo_Comb_type_ncRNA","Comb_type_exon_Oppo")]
} else {
  Peaksdata3_anno=Peaksdata2_anno[,c("ID",'chr',"start","end","strand",
                                'Counts_Unique','Counts_fracMM',
                                "Same_gene_name_comb","Same_ensembl_gene_id","Same_type_comb","Same_Repeat_comb",'Same_feature',"Same_Comb_type_exon","Same_Comb_type_ncRNA","Oppo_gene_name_comb","Oppo_ensembl_gene_id","Oppo_type_comb","Oppo_Repeat_comb",'Oppo_feature',"Oppo_Comb_type_exon","Oppo_Comb_type_ncRNA","Comb_type_exon_Oppo")]
}
Peaksdata3_anno=rename(Peaksdata3_anno,
Counts_Unique='Counts_Unique',
Counts_Multimappers_Scaled='Counts_fracMM',
`Same Strand: Host_gene_name`="Same_gene_name_comb",
`Same Strand: Host_gene_ensembl_id`="Same_ensembl_gene_id",
`Same Strand: RNA_type`="Same_type_comb",
`Same Strand: Repeat_Type`="Same_Repeat_comb",
`Same Strand: Intron Exon`='Same_feature',
`Same Strand: RNA_type_Classification`="Same_Comb_type_exon",
`Same Strand: ncRNA_Classification`="Same_Comb_type_ncRNA",
`Opposite Strand: Host_gene_name`="Oppo_gene_name_comb",
`Opposite Strand: Host_gene_ensembl_id`="Oppo_ensembl_gene_id",
`Opposite Strand: RNA_type`="Oppo_type_comb",
`Opposite Strand: Repeat_Type`="Oppo_Repeat_comb",
`Opposite Strand: Intron Exon`='Oppo_feature',
`Opposite Strand: RNA_type_Classification`="Oppo_Comb_type_exon",
`Opposite Strand: ncRNA_Classification`="Oppo_Comb_type_ncRNA",
`Both Strand: RNA_type_Classification`="Comb_type_exon_Oppo"
)
```

<br>  


# Output Data  


```{r ,fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T,include=F}
write.table(Peaksdata3_anno, file=output_table, sep = "\t", row.names = FALSE, col.names = T, append = F, quote= FALSE,na = "")
```
## Counts Table : 

"`r output_table`"  

**Columns:**  
"ID" - CLIP peak location  
"IDmerge" - CLIP regionns connected by spliced reads  
"strand" - CLIP peak stand  

"Counts_Unique" - Number of Unique reads aligned to CLIP peak  
"Counts_Multimappers_Scaled"  - number of reads with Unique reads =1 and Multimapped =1/#MM locations  

"Same Strand: Host_gene_name" - Name of Host Gene on same strand as CLIP reads  
"Same Strand: Host_gene_ensembl_id" - Ensemble Name of Host Gene on same strand as CLIP reads  
"Same Strand: RNA_type" - Transcript type of CLIP peak host gene on same strand as CLIP reads  
"Same Strand: Repeat_Type" - Type of repeat in Clip peak (determined by repeat masker) on same strand as CLIP reads  
"Same Strand: Intron Exon" - Gene feature of CLIP peak location in protein coding Gene on same strand as CLIP reads  
"Same Strand: RNA_type_Classification" - Summary of Peak Classification   
"Same Strand: ncRNA_Classification" - Summary of ncRNA Peak assignments  

"Opposite Strand: Host_gene_name" - Name of Host Gene on opposite strand as CLIP reads  
"Opposite Strand: Host_gene_ensembl_id" - Ensemble Name of Host Gene on opposite strand as CLIP reads  
"Opposite Strand: RNA_type" - Transcript type of CLIP peak host gene on opposite strand as CLIP reads  
"Opposite Strand: Repeat_Type" - Type of repeat in Clip peak (determined by repeat masker) on opposite strand as CLIP reads  
"Opposite Strand: Intron Exon" - Gene feature of CLIP peak location in protein coding Gene on opposite strand as CLIP reads  
"Opposite Strand: RNA_type_Classification" - Summary of Peak Classification  
"Opposite Strand: ncRNA_Classification" - Summary of ncRNA Peak assignments  

"Both Strand: RNA_type_Classification" - Summary of Peak Classification including both strand annotations 