---
title: ""
output: html_document
params:
  samplename: "WT1"
  peak_in: "/Users/homanpj/OneDrive - National Institutes of Health/Loaner/Wolin/CLIP/iCLIP_Git/workflow/scripts/AnnoTestPIPE/15_peaks/WT1_fCLIP_50nt_peakannotation_complete.txt"
  OutDIR: "/Users/homanpj/OneDrive - National Institutes of Health/Loaner/Wolin/CLIP/iCLIP_Git/workflow/scripts/AnnoTestPIPE/15_peaks/"
  ntmerge: "50nt"
  readdepth: 5
  species: 'hg38'
  Condense: FALSE
  JoinJunc: TRUE
  PeakIdnt: "all" #MultiMap, Unique, all
  NameAdd: "" ##"_SpliceAware_FullGenome_Transcme.SplicTrans_CollapseExons" ###need to remove this - we dont need additional id's
# params:
#   samplename: samplename
#   peak_in: peak_in
#   OutDIR: OutDIR
#   readdepth: readdepth
#   species: species
#   PeakIdnt: PeakIdnt # Unique, all
#   ntmerge: "50nt"
#   Condense: FALSE
#   JoinJunc: TRUE
#   calcMAPQ: FALSE
#   calcALTLOC: FALSE
#   DEmethod: DEmethod
#   RunMAnorm: FALSE
#   NameAdd: ""
---



```{r echo=F, message=FALSE, warning=FALSE, eval=F, include=F,results=F,}

if (params$PeakIdnt=='Unique') {peaksout='/uniquereadpeaks/'}
if (params$PeakIdnt=='all') {peaksout='/allreadpeaks/'}


rmarkdown::render('CLIP_MD16_BWpipe.Rmd',
                        encoding=encoding,
                        output_file=paste0(gsub('\\\\ ',' ',params$OutDIR),peaksout, params$samplename,'_CLIP_',params$PeakIdnt,'Peaks',params$ntmerge,'_peakDepth',params$readdepth,params$NameAdd,'_MD16.html'))
```
 
```{r envirment, include=F ,echo=F,warning=F,message=FALSE}
rm(list=setdiff(ls(), "params"))

library(rmarkdown)
library(stringr)
library(rtracklayer)
library(VariantAnnotation,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
# library(trackViewer)
library(rmarkdown)
# library(vcfR)
library(ggplot2,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library("viridis",quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(edgeR,quietly = T,verbose = F)
library('GenomicFeatures',quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library('rtracklayer',quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(matrixStats,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(plyr,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(tidyr,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(fitdistrplus,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(stringr,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(data.table)
library(reshape)
library(knitr)
library(stringi)
library(BSgenome)
library(biomaRt)
library(plotly)
library(tidyr)
library(GenomicRanges)
library(RColorBrewer)
library('gplots')
library(ggpubr)
library(circlize)
# library("Rgb")
library('regioneR')
 

blank_theme <- theme_minimal()+
  theme(
  axis.title.x = element_blank(),
  axis.title.y = element_blank(),
  panel.border = element_blank(),
  panel.grid=element_blank(),
  axis.ticks = element_blank(),
  plot.title=element_text(size=14, face="bold")
  )

incd_rRNA=T

```



```{r echo=F, message=FALSE, warning=FALSE, eval=T, include=F,results=F,}


peak_in=params$peak_in
samplename=params$samplename
INbam=params$INbam
INbamMM=params$INbamMM
INpeaks=params$INpeaks
inMAPQ=params$inMAPQ
ntmerge=params$ntmerge
readdepth=as.numeric(params$readdepth)
NameAdd=params$NameAdd
#calcMAPQ=params$calcMAPQ ###not used
#calcALTLOC=params$calcALTLOC ###not used
#RunMAnorm=params$calcALTLOC ###not used
species=params$species
OutDIR=params$OutDIR
PeakIdnt=params$PeakIdnt


if (params$PeakIdnt=='Unique') {peaksout='/uniquereadpeaks/'}
if (params$PeakIdnt=='all') {peaksout='/allreadpeaks/'}  

  if (dir.exists(file.path(paste0(OutDIR)))==F) {dir.create(paste0(OutDIR))}
  if (dir.exists(file.path(paste0(OutDIR,"/",peaksout)))==F) {dir.create(paste0(OutDIR,"/",peaksout))}


OutDIR=gsub("\\\\ "," ",OutDIR)

if (PeakIdnt=='Unique') {peaksout='/uniquereadpeaks/'}
if (PeakIdnt=='all') {peaksout='/allreadpeaks/'}

```

```{r echo=F, message=FALSE, warning=FALSE, eval=T, include=F,results=F,}

# Identification of CLIP peaks: `r toString(params$samplename)`    

 
Peaksdata2_anno = read.table(peak_in)

# unique(Peaksdata2_anno$Same_Comb_type_exon)
Peaksdata2_anno$Same_Comb_type_exon = 
    factor(Peaksdata2_anno$Same_Comb_type_exon, 
           levels = c("ncRNA", "protein_coding: exon", "Repeat Element","pseudogene",
                      "Antisense Feature","protein_coding: Intron","lncRNA","no Feature"))  

# unique(Peaksdata2_anno$Oppo_Comb_type_exon)
Peaksdata2_anno$Oppo_Comb_type_exon = 
    factor(Peaksdata2_anno$Oppo_Comb_type_exon, 
           levels = c("ncRNA", "protein_coding: exon", "Repeat Element","pseudogene",
                      "Antisense Feature","protein_coding: Intron","lncRNA","no Feature"))  

# unique(Peaksdata2_anno$Comb_type_exon_Oppo)
Peaksdata2_anno$Comb_type_exon_Oppo = 
    factor(Peaksdata2_anno$Comb_type_exon_Oppo, 
           levels = c("ncRNA", "protein_coding: exon", "Repeat Element","pseudogene",
                      "Antisense Feature","protein_coding: Intron","lncRNA","no Feature"))  

incd_rRNA=T 

```


  
<!-- ## Processing data    -->

**1. Align reads Novoalign**

<!-- novoalign -c 32 -t 15,3 -l 20 -x 4 -g 20 -s 1 -o SAM -R 0 -r All   -->
<!-- -d Hg38_gencode    -->
<!-- -f `r toString(samplename)`_Clip_iCountcutadpt.fastq   -->
<!-- <br>   -->

**2. UMI-tools remove duplicates**
<!-- umi_tools dedup   -->
<!-- --method unique   -->
<!-- --multimapping-detection-method=NH   -->
<!-- --umi-separator rbc:    -->
<!-- --log2stderr    -->
<!-- -I `r toString(samplename)`_Clip_iCountcutadpt_all.unique.NH.mm.s.bam     -->
<!-- -S `r toString(samplename)`_Clip_iCountcutadpt_all.unique.NH.mm.ddup.bam    -->
<!-- -L `r toString(samplename)`_Clip_iCountcutadpt_all.unique.NH.mm.s.log   -->
<!-- <br>   -->

**3. Select CLIP peaks**

  Identify peaks using bedtools merge on All reads (unique and Multimapping) and count number of reads that overlap with the identified peaks.
  
  Peaks with > `r toString(readdepth)` (unique + Multi-Mapped:fraction) reads were assessed.  


```{r ,fig.align='center',echo=F,warning=F,eval=T,include=T}



if (params$PeakIdnt=='Unique'){

x=ggplot(Peaksdata2_anno,aes(x=Counts_Unique))+geom_histogram(bins = 30)+scale_x_continuous(trans = "log2") + theme_classic()+ylab("# of peaks")+xlab("Peak Counts (unique)")+ggtitle("Distribution of peak counts")
#+geom_vline(xintercept = (100),color="red")+geom_vline(xintercept = (10),color="blue"))
ggplotly(x,tooltip = c("x", "y"))
}

 
if (params$PeakIdnt=='all'){

x=ggplot(Peaksdata2_anno,aes(x=Counts_fracMM))+geom_histogram(bins = 30)+scale_x_continuous(trans = "log2") + theme_classic()+ylab("# of peaks")+xlab("Peak Counts (Unique + FracMM)")+ggtitle("Distribution of peak counts")
#+geom_vline(xintercept = (100),color="red")+geom_vline(xintercept = (10),color="blue"))
ggplotly(x,tooltip = c("x", "y"))

}
```


<br/>    
<br/>    


## Identify CLIP peak Gene Location  

### Identify if CLIP peak overlaps with Intron or Exonic region   

<br>  

<!-- Intron coordinates were calculated from GTF file. -->

<!-- A second column was added to idenify if the peak also overlapped with the 5'UTR 3'UTR or CDS (Column: Featrue 2) -->

Protein Coding Genes were taken from `r refGTF` GTF file.     

Peaks were annotated with overlapping gene/transcript  

Peaks were annotated by whether they overlap with Host gene intron/exon region  

<br>   

### Identify CLIP peak in ncRNA   


```{r echo=F,results='asis',eval=F}
Classification_ncRNA=read.delim(paste0(OUT,"ncRNA_Annotations.txt"), header=T, sep="\t",stringsAsFactors = F)

# Classification_ncRNA['rRNA_DNA','contents']=gsub(c(', TermrRNA, repeat_RI-1, repeat_RII-1, APS1_X52413_1, APS1_X52413_2, Block_1, repeat_RI-2, repeat_RII-2, APS1_X52413_3, APS2_X52412_1, Block_2, repeat_RI-3, repeat_RII-3, APS1_X52413_4, repeat_RII-4, APS2_52412_2, Sal_box_1, Sal_box_2, Sal_box_3, Sal_box_4, Sal_box_5, Sal_box_6, Sal_box_7, Sal_box_8, Sal_box_9, spacer_promoter_X06186, enhancer_V00847, Sal_box_0, promoter_M18062, UPE, core_rRNA_promoter'),"",Classification_ncRNA['rRNA_DNA','contents'])

kable(Classification_ncRNA[,1:3])

```

<br>  


### Identify peaks in repeat regions   

   Annotate all repeat regions/Classes identified in Repeatmasker Annotation file (UCSC Table browser)  
   Data was not filtered based on any of the identified Repeats.  
    1) LINE/SINE   
    2) LTR   
    3) DNA   
    4) Satellites   
    5) Simple Repeats   
    6) Low Complexity   
    7) Other   
    8) Unknown   
    <br>   


```{r echo=F,warning=F, include=T,eval=T,fig.align='center'}

p=as.data.frame(((colSums(is.na(Peaksdata2_anno[,grep('Same_Repeat_',colnames(Peaksdata2_anno))])==F)/nrow(Peaksdata2_anno))*100))
p$Repeat_Type=rownames(p);colnames(p)=c('Frequency','Repeat_Type')
p=p[rownames(p)%in%'Same_Repeat_comb'==F,]
p$Repeat_Type=gsub("Same_Repeat_","",p$Repeat_Type)

ggplot(p )+geom_bar(aes(x=Repeat_Type,y=Frequency),stat="identity")+theme_classic()+#ggtitle("SNP Mutations")+
  xlab("")+ylab("Frequency of peaks in \nrepeat regions (%)")+ggtitle("Same Strand")+
theme(plot.title = element_text(hjust = 0.5,size = 15),axis.title=element_text(size=15),axis.text=element_text(size=10),legend.text = element_text( size=10))+theme(axis.text.x = element_text(angle = -45))
```
   
<br>   
<br>   

### Asigning Clip peak attributes   

Not all Peaks overlap with a single feature so peak assignments were assigned by priority:  

<!-- **ncRNA > Protein coding : Exonic > repeats > Pseudogene > Antisense Feature > Protein Coding : Intronic > lncRNA > no Feature**    -->

**ncRNA > protein_coding: Exonic > Repeat Element > pseudogene > lncRNA: Exonic > Antisense Feature > protein_coding: Intronic > lncRNA: Intronic > no Feature**

<!-- rDNA transcript (BK000964.3) was included in reference genome but not included in peak counts. rDNA will affect multimapping count and will show up in alternative mapping location.   	 -->

All annotations from RNA type, Repeat regions, and Intronic/exonic regions are annotated in the Table.   

<br>  
<br>  


## Overview of Ro targets by CLIP peak catagory   

```{r , echo=F, include=T, eval=T, fig.align='default', fig.ncol=2, fig.height=3.5, fig.width=7, out.height=200, out.width="50%", fig.show="hold", layout='l-screen', results = FALSE, message=FALSE , warning=F}
if (params$PeakIdnt=='Unique'){
  x1=ggplot(Peaksdata2_anno,aes(x=Counts_Unique,y=Length,color=Comb_type_exon_Oppo,text=Same_external_gene_name))+geom_point(size=1)+
    scale_x_continuous(trans = "log2",limits = c(.5,NA))+
    scale_y_continuous(trans = "log2",limits = c(NA,NA)) +
    theme_classic()+
  ylab("Peak Width")+xlab("Peak Counts (unique)")+ggtitle("Comparison of peak counts and Peak Width")+
  scale_color_brewer(palette = "Dark2") 
# ggplotly(x1,tooltip = c("x", "y","Same_external_gene_name"))

}
if (params$PeakIdnt=='MultiMap'){
x1=ggplot(Peaksdata2_anno,aes(x=Counts_MM,y=Length,color=Comb_type_exon_Oppo,text=Same_external_gene_name))+geom_point(size=1)+
    scale_x_continuous(trans = "log2",limits = c(4,NA))+
    scale_y_continuous(trans = "log2",limits = c(NA,NA))  + 
  theme_classic()+
  ylab("Peak Width")+xlab("Peak Counts (MultiMap)")+ggtitle("Comparison of peak counts and Peak Width")+
  scale_color_brewer(palette = "Dark2") 
# ggplotly(x1,tooltip = c("x", "y","Same_external_gene_name"))
}

if (params$PeakIdnt=='all'){
x1=ggplot(Peaksdata2_anno,aes(x=Counts_fracMM,y=Length,color=Comb_type_exon_Oppo,text=Same_external_gene_name))+geom_point(size=1)+
    scale_x_continuous(trans = "log2",limits = c(4,NA))+
    scale_y_continuous(trans = "log2",limits = c(NA,NA)) + 
  theme_classic()+
  ylab("Peak Width")+xlab("Peak Counts (Unique+ FracMM)")+ggtitle("Comparison of peak counts and Peak Width")+  labs(color='RNA Type')+
  scale_color_brewer(palette = "Dark2") 
# ggplotly(x1,tooltip = c("x", "y","Same_external_gene_name"))
}

x2=ggplot(Peaksdata2_anno,aes(x=Length))+geom_histogram(bins = 100)+
  scale_x_continuous(trans = "log2",limits = c(20,NA)) + 
  theme_classic()+ylab("# of peaks")+xlab("Peak Width")+ggtitle("Distribution of peak Width")
# ggplotly(x2,tooltip = c("x", "y"))

x3=ggplot(Peaksdata2_anno,aes(x=Length,line=Comb_type_exon_Oppo,color=Comb_type_exon_Oppo))+geom_density(size=.5)+
  scale_x_continuous(trans = "log2",limits = c(8,NA)) + 
  theme_classic()+ylab("Density")+xlab("Peak Width")+ggtitle("Distribution of peak Width")+  labs(color='RNA Type')+
  scale_color_brewer(palette = "Dark2") 
# ggplotly(x3,tooltip = c("x", "y","line"))


x4=ggplot(Peaksdata2_anno,aes(x=Length,line=Comb_type_exon_Oppo,color=Comb_type_exon_Oppo))+geom_freqpoly(bins = 100)+
  scale_x_continuous(trans = "log2",limits = c(8,NA)) + 
  theme_classic()+ylab("# of peaks")+xlab("Peak Width")+ggtitle("Distribution of peak Width")+  labs(color='RNA Type')+
  scale_color_brewer(palette = "Dark2") 
# ggplotly(x4,tooltip = c("x", "y","line"))


# x5=ggplot(Peaksdata2_anno,aes(x=Counts_Unique,line=Comb_type_exon_Oppo,color=Comb_type_exon_Oppo))+geom_freqpoly(bins = 100)+scale_x_continuous(trans = "log2") + theme_classic()+ylab("# of peaks")+xlab("Peak Counts (unique)")+ggtitle("Distribution of peak counts")+  labs(color='RNA Type')+
#   scale_color_brewer(palette = "Dark2") 
# # ggplotly(x4,tooltip = c("x", "y","line"))
if (params$PeakIdnt=='Unique'){

x5=ggplot(Peaksdata2_anno,aes(x=Counts_Unique,line=Comb_type_exon_Oppo,color=Comb_type_exon_Oppo))+geom_density()+
  scale_x_continuous(trans = "log2",limits = c(.25,NA)) + 
  theme_classic()+ylab("Density")+xlab("Peak Counts (unique)")+ggtitle("Distribution of peak counts")+  labs(color='RNA Type')+
  scale_color_brewer(palette = "Dark2") 
# ggplotly(x4,tooltip = c("x", "y","line"))
}
if (params$PeakIdnt=='MultiMap'){

  x5=ggplot(Peaksdata2_anno,aes(x=Counts_Unique,line=Comb_type_exon_Oppo,color=Comb_type_exon_Oppo))+geom_density()+
    scale_x_continuous(trans = "log2",limits = c(.25,NA)) + 
    theme_classic()+ylab("Density")+xlab("Peak Counts (Counts_MM)")+ggtitle("Distribution of peak counts")+  labs(color='RNA Type')+
  scale_color_brewer(palette = "Dark2") 
# ggplotly(x4,tooltip = c("x", "y","line"))
}
if (params$PeakIdnt=='all'){

  x5=ggplot(Peaksdata2_anno,aes(x=Counts_Unique,line=Comb_type_exon_Oppo,color=Comb_type_exon_Oppo))+geom_density()+
    scale_x_continuous(trans = "log2",limits = c(.25,NA)) + 
    theme_classic()+ylab("Density")+xlab("Peak Counts (Unique+ FracMM)")+ggtitle("Distribution of peak counts")+  labs(color='RNA Type')+
  scale_color_brewer(palette = "Dark2") 
# ggplotly(x4,tooltip = c("x", "y","line"))
}
# subplot(style(x3, showlegend = F,showtitle=F,margin = .5),
        # style(x4, showlegend = T,showtitle=F,margin = .5),
        # shareX=T,shareY=F,titleX=T,titleY=T,heights=c(.5),widths = c(.5,.5))
x1;x2;x3;x4;x5
# 
# library(gridExtra)
# grid.arrange(x3, x4, ncol=2)

```

<br>  
<br>   

<!-- ##  Peak Attributes    -->

##  Number of peaks assigned to each catagory  


**Peak counts do include rRNA peaks**   

```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,eval=T,include=T}
p=Peaksdata2_anno
p$Same_Comb_type_exon=p$Comb_type_exon_Oppo

if (incd_rRNA==F) {
  p=p[!p$Same_Comb_type_ncRNA%in%'rRNA',]
}

p[is.na(p$Same_Comb_type_exon),'Same_Comb_type_exon']='no Feature'
colnames(p)[colnames(p)%in%'Same_Comb_type_exon']='RNA_Type'
p$RNA=''
gg=ggplot(p,aes(fill=RNA_Type,x=RNA) )+geom_bar(position="stack",width=.5)+theme_classic()+ggtitle("BioType - Counted by catagory\n(Each Clip Peak only Counted Once)")+
# theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15),legend.text = element_text( size=30))+
 scale_fill_brewer(palette = "Dark2") 
ggplotly(gg)

```

<!-- <font size="1">Columns: RNA_type_Classification </font>      -->
<!-- <br/>      -->

```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,eval=T,include=T}


p=Peaksdata2_anno
p$Same_Comb_type_exon=p$Comb_type_exon_Oppo
if (incd_rRNA==F) {
  p=p[!p$Same_Comb_type_ncRNA%in%'rRNA',]
}

p=p[p$Same_Comb_type_exon%in%'ncRNA','Same_Comb_type_ncRNA',drop=F]
if (nrow(p)>0) {
  
p$RNA=""
gg=ggplot(p,aes(fill=Same_Comb_type_ncRNA,x=RNA) )+geom_bar(width=.5)+theme_classic()+ggtitle("ncRNA only: # of peaks per catagory")+  labs(fill='RNA Type')+
 scale_color_brewer(palette = "Paired")+scale_fill_brewer(palette = "Paired") +  blank_theme+ 
  coord_polar("y",start=0)
(gg)

}

 
```    

<br>  
<br> 

**Peak counts do not include rRNA peaks**   


 
```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,eval=T,include=T}


p=Peaksdata2_anno
p$Same_Comb_type_exon=p$Comb_type_exon_Oppo

  p=p[!p$Same_Comb_type_ncRNA%in%'rRNA',]


p[is.na(p$Comb_type_exon),'Same_Comb_type_exon']='no Feature'
colnames(p)[colnames(p)%in%'Same_Comb_type_exon']='RNA_Type'
p$RNA=''
gg=ggplot(p,aes(fill=RNA_Type,x=RNA) )+geom_bar(position="stack",width=.5)+theme_classic()+ggtitle("Number of peaks by catagory")+
# gg=ggplot(p,aes(fill=RNA_Type,x=RNA) )+geom_bar(position="stack",width=.5)+theme_classic()+ggtitle("Number of peaks by catagory\n(Determined by unique reads only)")+
# theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15),legend.text = element_text( size=30))+
 scale_fill_brewer(palette = "Dark2") 
ggplotly(gg)



```


<!-- <font size="1">Columns: RNA_type_Classification </font>      -->
<!-- <br/>      -->


```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,eval=T,include=T}


p=Peaksdata2_anno


p$Same_Comb_type_exon=p$Comb_type_exon_Oppo
# if (incd_rRNA==F) {
  p=p[!p$Same_Comb_type_ncRNA%in%'rRNA',]
# }

p=p[p$Same_Comb_type_exon%in%'ncRNA','Same_Comb_type_ncRNA',drop=F]
if (nrow(p)>0) {
  

p$RNA=""
gg=ggplot(p,aes(fill=Same_Comb_type_ncRNA,x=RNA) )+geom_bar(width=.5)+theme_classic()+
  ggtitle("Number of peaks by catagory : ncRNA only\n(Determined by unique reads only)")+  labs(fill='ncRNA Type')+
 scale_color_brewer(palette = "Paired")+ 
  scale_fill_brewer(palette = "Paired") +  blank_theme+ 
  coord_polar("y",start=0)
(gg)
}

# colnames(Peaksdata3)
```       

  

<br>  
<br>   
<br>   

## Number of reads assigned to each catagory   

#### Counts for RNA biotype  

<br>   

```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,include=T,eval=T}
# brewer.pal(n = 8, name = "Dark2")

nc_color=c("#1B9E77","#1fb487",
           "#2bdba6" ,
           "#D95F02", "#7570B3" ,"#E7298A", "#66A61E", "#E6AB02", "#A6761D", "#666666")

p=Peaksdata2_anno
p$Same_Comb_type_exon=p$Comb_type_exon_Oppo
# if (incd_rRNA==F) {
#   p=p[!p$Same_Comb_type_ncRNA%in%'rRNA',]
# }
p$Same_Comb_type_exon=as.character(p$Same_Comb_type_exon)
p[p$Same_Comb_type_ncRNA%in%'rRNA','Same_Comb_type_exon']='rRNA'
p[p$Same_Comb_type_ncRNA%in%'yRNA','Same_Comb_type_exon']='yRNA'

p$Same_Comb_type_exon=factor(p$Same_Comb_type_exon, levels = c("ncRNA","rRNA","yRNA", "protein_coding: exon", "Repeat Element","pseudogene","Antisense Feature","protein_coding: Intron","lncRNA","no Feature"))
# p$Same_Comb_type_exon=factor(p$Same_Comb_type_exon, levels = c("ncRNA","rRNA", "protein_coding: exon", "Repeat Element","pseudogene","Antisense Feature","protein_coding: Intron","lncRNA","no Feature"))


# u=unique(p$Same_Comb_type_exon) ##Change 05/04
u=c("ncRNA","rRNA","yRNA", "protein_coding: exon", "Repeat Element","pseudogene","Antisense Feature","protein_coding: Intron","lncRNA","no Feature")
pcount=data.frame(u,1:length(u));colnames(pcount)=c('RNA_Type','Counts_All');pcount$Counts_Unique=NA;pcount$Counts_MM=NA;pcount$Peak_count=NA
for (x in 1:length(u)) {
  pam=p[p$Same_Comb_type_exon%in%u[x],]
  pcount[x,1]=u[x]
  # pcount[x,'Counts_All']=colSums(pam[,'Counts_All',drop=F],na.rm = T)
  pcount[x,'Counts_Unique']=colSums(pam[,'Counts_Unique',drop=F],na.rm = T)
  # pcount[x,'Counts_MM']=colSums(pam[,'Counts_MM',drop=F],na.rm = T)
  # pcount[x,'Counts_primary']=colSums(pam[,'Counts_primary',drop=F],na.rm = T)
  pcount[x,'Counts_fracMM']=colSums(pam[,'Counts_fracMM',drop=F],na.rm = T)
  # pcount[x,'Counts_Primary_fracMM']=colSums(pam[,'Counts_Primary_fracMM',drop=F],na.rm = T)
    pcount[x,'Peak_count']=nrow(pam)

}
pcount$RNA=""
pcount$RNA_Type=factor(pcount$RNA_Type, levels = c("ncRNA","rRNA","yRNA", "protein_coding: exon", "Repeat Element","pseudogene","Antisense Feature","protein_coding: Intron","lncRNA","no Feature"))



if (params$PeakIdnt=='MultiMap'){

gg=
ggplot(pcount,aes(fill=RNA_Type,x=RNA,y=Counts_MM) )+geom_bar(stat='identity',position="stack",width=.5)+theme_classic()+ggtitle("BioType - # of reads (MultiMap) by catagory)")+
 scale_fill_manual(values=nc_color)
  # scale_fill_brewer(palette = "Dark2") 
  
ggplotly(gg)
}

# if (params$PeakIdnt=='Unique'|params$PeakIdnt=='all'){

gg=
ggplot(pcount,aes(fill=RNA_Type,x=RNA,y=Counts_Unique) )+geom_bar(stat='identity',position="stack",width=.5)+theme_classic()+ggtitle("BioType - # of reads (unique) by catagory)")+
 scale_fill_manual(values=nc_color)
  # scale_fill_brewer(palette = "Dark2") 
  
ggplotly(gg)
# }



gg=
ggplot(pcount,aes(fill=RNA_Type,x=RNA,y=Counts_fracMM) )+geom_bar(stat='identity',position="stack",width=.5)+theme_classic()+ggtitle("BioType - # of reads (unique + Frac. MM) by catagory)")+
scale_fill_manual(values=nc_color)
  # scale_fill_brewer(palette = "Dark2")
ggplotly(gg)


```
<br>    

#### Counts for ncRNA      


```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,include=T,eval=T}

p=Peaksdata2_anno
p$Same_Comb_type_exon=p$Comb_type_exon_Oppo
# if (incd_rRNA==F) {
#   p=p[!p$Same_Comb_type_ncRNA%in%'rRNA',]
# }
p$Same_Comb_type_exon=as.character(p$Same_Comb_type_exon)
p[p$Same_Comb_type_ncRNA%in%'rRNA','Same_Comb_type_exon']='rRNA'
p[p$Same_Comb_type_ncRNA%in%'yRNA','Same_Comb_type_exon']='yRNA'

p$Same_Comb_type_exon=factor(p$Same_Comb_type_exon, levels = c("ncRNA","rRNA","yRNA", "protein_coding: exon", "Repeat Element","pseudogene","Antisense Feature","protein_coding: Intron","lncRNA","no Feature"))
# p$Same_Comb_type_exon=factor(p$Same_Comb_type_exon, levels = c("ncRNA","rRNA", "protein_coding: exon", "Repeat Element","pseudogene","Antisense Feature","protein_coding: Intron","lncRNA","no Feature"))

p=p[p$Same_Comb_type_exon%in%'ncRNA',]
if (nrow(p)>0) {
  
u=unique(p$Same_Comb_type_ncRNA)
pcount=data.frame(u,1:length(u));colnames(pcount)=c('Same_RNA_Subtype','Counts_All');pcount$Counts_Unique=NA;pcount$Counts_MM=NA;pcount$Peak_count=NA
for (x in 1:length(u)) {
  pam=p[p$Same_Comb_type_ncRNA%in%u[x],]
  pcount[x,1]=u[x]
  # pcount[x,'Counts_All']=colSums(pam[,'Counts_All',drop=F],na.rm = T)
  pcount[x,'Counts_Unique']=colSums(pam[,'Counts_Unique',drop=F],na.rm = T)
  # pcount[x,'Counts_MM']=colSums(pam[,'Counts_MM',drop=F],na.rm = T)
    # pcount[x,'Counts_primary']=colSums(pam[,'Counts_primary',drop=F],na.rm = T)
  pcount[x,'Counts_fracMM']=colSums(pam[,'Counts_fracMM',drop=F],na.rm = T)
  # pcount[x,'Counts_Primary_fracMM']=colSums(pam[,'Counts_Primary_fracMM',drop=F],na.rm = T)
  pcount[x,'Peak_count']=nrow(pam)
}

pcount$RNA=''



gg1=ggplot(pcount,aes(fill=Same_RNA_Subtype,x=RNA,y=Counts_Unique) )+geom_bar(stat='identity',width=.5)+theme_classic()+ggtitle("ncRNA only: # of reads (unique) by catagory")+  labs(color='RNA Type')+
 scale_color_brewer(palette = "Paired")+scale_fill_brewer(palette = "Paired") +  blank_theme+ 
  coord_polar("y",start=0)


gg2=ggplot(pcount,aes(fill=Same_RNA_Subtype,x=RNA,y=Counts_fracMM) )+geom_bar(stat='identity',width=.5)+theme_classic()+ggtitle("ncRNA only: # of reads (unique + Frac. MM) by catagory")+  labs(color='RNA Type')+
 scale_color_brewer(palette = "Paired")+scale_fill_brewer(palette = "Paired") +  blank_theme+ 
  coord_polar("y",start=0)
plot(gg1)
plot(gg2)
}

```      
<!-- <font size="1">Columns: RNA_type_Classification + Counts_Unique </font>      -->

<br>   

<br>   

### Read Counts by peak for each catagory    
  
#### Unique Read counts (Count Unique reads only)   

<br>   

```{r fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T, include=T}
p=Peaksdata2_anno
p$Same_Comb_type_exon=p$Comb_type_exon_Oppo
if (incd_rRNA==F) {
  p=p[!p$Same_Comb_type_ncRNA%in%'rRNA',]
}

colnames(p)[colnames(p)%in%'Same_Comb_type_exon']='RNA_Type'
p$RNA=''
 gg=ggplot(p,aes(x=RNA_Type,y=Counts_Unique,fill=RNA_Type) )+geom_violin()+theme_classic()+geom_jitter(height = 0, width = 0.1,size=1)+
  scale_y_continuous(trans = "log2")+
  ggtitle('Peak Counts (Unique) for Ro-Clip\nby Feature')+ylab("Counts - Ro-Clip")+xlab("")+
theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15))+theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle = -45))
# 
gg
```  
<!-- <font size="1">Columns: RNA_type_Classification </font>      -->
<br/>    

<br/>    

```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=8,echo=F,warning=F,eval=T, include=T}
p=Peaksdata2_anno
p$Same_Comb_type_exon=p$Comb_type_exon_Oppo
# if (incd_rRNA==F) {
#   p=p[!p$Same_Comb_type_ncRNA%in%'rRNA',]
# }

p=p[p$Same_Comb_type_exon%in%'ncRNA',]

if (nrow(p)>0) {

p$ID_name=paste0(p$ID,"\n",p$Same_gene_name_comb)


colnames(p)[colnames(p)%in%'Same_Comb_type_ncRNA']='RNA_Type'


 gg=ggplot(p,aes(x=RNA_Type,y=Counts_Unique,fill=RNA_Type) )+geom_violin()+theme_classic()+geom_jitter(height = 0, width = 0.1,size=0.5,aes(text=ID_name))+
  scale_y_continuous(trans = "log2")+
  ggtitle('Peak Counts (Unique) for Ro-Clip\nby Feature')+ylab("Counts - Ro-Clip")+xlab("")+
theme(plot.title = element_text(hjust = 0.5,size = 15),axis.title=element_text(size=15),axis.text=element_text(size=12))+theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle = -45))+scale_fill_brewer(palette = "Paired")+ylim(min(p$Counts_Unique),max(p$Counts_Unique))
 ggplotly(gg,tooltip = c("text", "y"))
}
```  
<br>   
<br>   

#### Fractional Multimap Read counts (Count Unique reads + (MultiMapped read / # of different mappings) )     

Does include rRNA
<br>   

```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T}
p=Peaksdata2_anno
p$Same_Comb_type_exon=p$Comb_type_exon_Oppo
# if (incd_rRNA==F) {
#   p=p[!p$Same_Comb_type_ncRNA%in%'rRNA',]
# }

colnames(p)[colnames(p)%in%'Same_Comb_type_exon']='RNA_Type'
p$RNA=''
 gg=ggplot(p,aes(x=RNA_Type,y=Counts_fracMM,fill=RNA_Type) )+geom_violin()+theme_classic()+geom_jitter(height = 0, width = 0.1,size=1)+
  scale_y_continuous(trans = "log2")+
  ggtitle('Peak Counts (Unique + Frac. MM) for Ro-Clip\nby Feature')+ylab("Counts - Ro-Clip")+xlab("")+
theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15))+theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle = -45))
# 
gg
```  
<!-- <font size="1">Columns: RNA_type_Classification </font>      -->
<br/>    

<br/>    

```{r fig.asp=.5,fig.align='center',fig.height=6, fig.width=8,echo=F,warning=F,eval=T}
p=Peaksdata2_anno
p$Same_Comb_type_exon=p$Comb_type_exon_Oppo
# if (incd_rRNA==F) {
#   p=p[!p$Same_Comb_type_ncRNA%in%'rRNA',]
# }

p=p[p$Same_Comb_type_exon%in%'ncRNA',]
if (nrow(p)>0) {

colnames(p)[colnames(p)%in%'Same_Comb_type_ncRNA']='RNA_Type'
p$ID_name=paste0(p$ID,"\n",p$Same_gene_name_comb)

 gg=ggplot(p,aes(x=RNA_Type,y=Counts_fracMM,fill=RNA_Type) )+geom_violin()+theme_classic()+geom_jitter(height = 0, width = 0.05,size=.5,aes(text=ID_name))+
  scale_y_continuous(trans = "log2")+
  ggtitle('Peak Counts (Unique + Frac. MM) for Ro-Clip\nby ncRNA Type')+ylab("Counts - Ro-Clip")+xlab("")+
theme(plot.title = element_text(hjust = 0.5,size = 15),axis.title=element_text(size=15),axis.text=element_text(size=12))+theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle = -45))+scale_fill_brewer(palette = "Paired")+ylim(min(p$Counts_Unique),max(p$Counts_Unique))

 ggplotly(gg,tooltip = c("text", "y"))
 
}
```  
<br>   
<br>   

<br>  


## Opposite strand CLIP peak attributes   


```{r , fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F}

p=Peaksdata2_anno
p=p[p$Comb_type_exon_Oppo%in%'Antisense Feature',]
if (incd_rRNA==F) {
  p=p[!p$Same_Comb_type_ncRNA%in%'rRNA',]
}

colnames(p)[colnames(p)%in%'Oppo_Comb_type_exon']='RNA_Type'
p$RNA=''

ggplot(p,aes(fill=RNA_Type,x=RNA_Type) )+geom_bar(stat="count")+theme_classic()+
  ggtitle("Opposite Strand Feature - Number of peaks by catagory\n(Determined by unique reads only)")+xlab("")+
theme(plot.title = element_text(hjust = 0.5),axis.title=element_text(size=15),axis.text=element_text(size=10),legend.position = "none",axis.text.x = element_text(angle = -45,vjust = .5))+
 scale_fill_brewer(palette = "Dark2") 


```

```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T,incude=T}

p=Peaksdata2_anno
if (incd_rRNA==F) {
  p=p[!p$Same_Comb_type_ncRNA%in%'rRNA',]
}

p$transition=paste0(p$Same_Comb_type_exon,'->\n',p$Oppo_Comb_type_exon)
p=p[p$Comb_type_exon_Oppo%in%'Antisense Feature',]
colnames(p)[colnames(p)%in%'transition']='RNA_Type'
p$Difference=''

ggplot(p,aes(fill=Same_Comb_type_exon,x=factor(Oppo_Comb_type_exon)) )+geom_bar(stat="count")+theme_classic()+
  ggtitle("Opposite Strand Feature Features: \n Compare Antisense and Snese stand Annotation")+xlab("Antisense strand Annotation")+ylab("Peak Count")+labs(fill = "Sense Stand Annotation")+
theme(plot.title = element_text(hjust = 0.5,size=20),axis.title=element_text(size=15),axis.text=element_text(size=10),legend.position = "right",axis.text.x = element_text(angle = -45,vjust = .5))


```

<br>  

```{r echo=F,warning=F, include=T,eval=T}

p=as.data.frame(((colSums(is.na(Peaksdata2_anno[,grep('Oppo_Repeat_',colnames(Peaksdata2_anno))])==F))))
p$Repeat_Type=rownames(p);colnames(p)=c('Frequency','Repeat_Type')
p=p[rownames(p)%in%'Oppo_Repeat_comb'==F,]
p$Repeat_Type=gsub("Oppo_Repeat_","",p$Repeat_Type)

ggplot(p )+geom_bar(aes(x=Repeat_Type,y=Frequency),stat="identity")+theme_classic()+#ggtitle("SNP Mutations")+
  xlab("")+ylab("number of peaks in \nrepeat regions")+ggtitle("Opposite Strand")+
theme(plot.title = element_text(hjust = 0.5,size = 15),axis.title=element_text(size=15),axis.text=element_text(size=10),legend.text = element_text( size=10))+theme(axis.text.x = element_text(angle = -45))

```

```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T,include=F}
#RUN2


if ('IDmerge'%in% colnames(Peaksdata2_anno)) {
  Peaksdata3_anno=Peaksdata2_anno[,c("ID","IDmerge",'chr',"start","end","strand",
                                'Counts_Unique','Counts_fracMM',
                                "Same_gene_name_comb","Same_ensembl_gene_id","Same_type_comb","Same_Repeat_comb",'Same_feature',"Same_Comb_type_exon","Same_Comb_type_ncRNA","Oppo_gene_name_comb","Oppo_ensembl_gene_id","Oppo_type_comb","Oppo_Repeat_comb",'Oppo_feature',"Oppo_Comb_type_exon","Oppo_Comb_type_ncRNA","Comb_type_exon_Oppo")]
} else {
  Peaksdata3_anno=Peaksdata2_anno[,c("ID",'chr',"start","end","strand",
                                'Counts_Unique','Counts_fracMM',
                                "Same_gene_name_comb","Same_ensembl_gene_id","Same_type_comb","Same_Repeat_comb",'Same_feature',"Same_Comb_type_exon","Same_Comb_type_ncRNA","Oppo_gene_name_comb","Oppo_ensembl_gene_id","Oppo_type_comb","Oppo_Repeat_comb",'Oppo_feature',"Oppo_Comb_type_exon","Oppo_Comb_type_ncRNA","Comb_type_exon_Oppo")]
}


Peaksdata3_anno=rename(Peaksdata3_anno,
Counts_Unique='Counts_Unique',
Counts_Multimappers_Scaled='Counts_fracMM',
`Same Strand: Host_gene_name`="Same_gene_name_comb",
`Same Strand: Host_gene_ensembl_id`="Same_ensembl_gene_id",
`Same Strand: RNA_type`="Same_type_comb",
`Same Strand: Repeat_Type`="Same_Repeat_comb",
`Same Strand: Intron Exon`='Same_feature',
`Same Strand: RNA_type_Classification`="Same_Comb_type_exon",
`Same Strand: ncRNA_Classification`="Same_Comb_type_ncRNA",
`Opposite Strand: Host_gene_name`="Oppo_gene_name_comb",
`Opposite Strand: Host_gene_ensembl_id`="Oppo_ensembl_gene_id",
`Opposite Strand: RNA_type`="Oppo_type_comb",
`Opposite Strand: Repeat_Type`="Oppo_Repeat_comb",
`Opposite Strand: Intron Exon`='Oppo_feature',
`Opposite Strand: RNA_type_Classification`="Oppo_Comb_type_exon",
`Opposite Strand: ncRNA_Classification`="Oppo_Comb_type_ncRNA",
`Both Strand: RNA_type_Classification`="Comb_type_exon_Oppo"
)




```

<br>  


# Output Data  


```{r ,fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T,include=F}


write.table(Peaksdata3_anno,file=paste0(OutDIR,peaksout,samplename,"_CLIP_",params$PeakIdnt,"Peaks",ntmerge,'_peakDepth',params$readdepth,params$NameAdd,"_MD16.txt"), sep = "\t", row.names = FALSE, col.names = T, append = F, quote= FALSE,na = "")
# colnames(Peaksdata4)

```


```{r echo=F, message=FALSE, warning=FALSE, eval=F, include=T,results=F,}

Peaksdata2_anno_OLD=read.table('/Users/homanpj/OneDrive - National Institutes of Health/Loaner/Wolin/CLIP/iCLIP_Git/test/14_peaks/allreadpeaks/WT1_CLIP_allPeaks50nt_peakDepth5_MD16.txt',sep = '\t',header = T)
# Peaksdata2_anno_OLD=read.table('/Users/homanpj/OneDrive - National Institutes of Health/Loaner/Wolin/CLIP/NextSeq_fCLIP_0218/14_peaks/allreadpeaks/WT1_CLIP_allPeaks50nt_peakDepth5_MD16.txt',sep = '\t',header = T)

Peaksdata2_anno_OLD=separate(Peaksdata2_anno_OLD,"ID",sep=":|-",into = c('chr','start','end'),remove = F)
Peaksdata2_anno_OLD$start=as.numeric(Peaksdata2_anno_OLD$start)-1
Peaksdata2_anno_OLD$end=as.numeric(Peaksdata2_anno_OLD$end)
Peaksdata2_anno_OLD$ID=paste0(Peaksdata2_anno_OLD$chr,':',Peaksdata2_anno_OLD$start,'-',Peaksdata2_anno_OLD$end)

setdiff(Peaksdata3_anno$ID,Peaksdata2_anno_OLD$ID)
setdiff(Peaksdata3_anno$ID,Peaksdata2_anno$ID)
# View(Peaksdata2_anno[Peaksdata2_anno$ID%in%setdiff(Peaksdata2_anno$ID,Peaksdata2_anno_OLD$ID),])
# View(Peaksdata2_anno_OLD[Peaksdata2_anno_OLD$ID%in%setdiff(Peaksdata2_anno_OLD$ID,Peaksdata2_anno$ID),])

Peaksdata3_anno_Comb=merge(Peaksdata3_anno,Peaksdata2_anno_OLD[,c('ID','Both.Strand..RNA_type_Classification')],by="ID")

test=Peaksdata3_anno_Comb[Peaksdata3_anno_Comb$`Both Strand: RNA_type_Classification` != Peaksdata3_anno_Comb$Both.Strand..RNA_type_Classification,]
test2=Peaksdata3_anno_Comb[Peaksdata3_anno_Comb$Comb_type_exon_Oppo  != Peaksdata3_anno_Comb$Both.Strand..RNA_type_Classification,]


comp1=(test2$Comb_type_exon_Oppo=="no Feature")
comp2=grepl('linLcRNA-exon',test2[,paste0('Same_','type_comb')]) | grepl('linLcRNA',test2[,paste0('Same_','type_comb')])
comp2=((test2[,paste0('Same_','type_comb')]%in%c('linLcRNA-exon')) | (test2[,paste0('Same_','type_comb')]%in%c('lnLcRNA')) )
comp=comp1&comp2
PeaksdataOut[comp,'Comb_type_exon_Oppo']='lnLcRNA-exon'
#


# Peaksdata2_anno_Comb=merge(Peaksdata2_anno,Peaksdata2_anno_OLD[,c('ID','Both.Strand..RNA_type_Classification')],by="ID")
# test2=Peaksdata2_anno_Comb[Peaksdata2_anno_Comb$Comb_type_exon_Oppo  != Peaksdata2_anno_Comb$Both.Strand..RNA_type_Classification,]

```


## Counts Table : 

"`r paste0(samplename,"_CLIP_",params$PeakIdnt,"Peaks",ntmerge,'_peakDepth',params$readdepth,params$NameAdd,"_MD16.txt")`"  

**Columns:**  
"ID" - CLIP peak location  
"strand" - CLIP peak stand  
<!-- "Average MAPQ" - Average Mapping quality of reads aligned to peak region.   -->
"Counts_Unique" - Number of Unique reads aligned to CLIP peak  
<!-- "Counts_All" - Count of Multimapped reads for each peak     -->
<!-- "Counts_All" - Count of Unique + Multimapped read for each peak (Each occurrence of Multimapped read is counted separately)   -->
"Counts_Multimappers_Scaled"  - number of reads with Unique reads =1 and Multimapped =1/#MM locations  
<!-- "Counts_Multimappers_Best Mapping" - Number of Unique reads + best mapQ Multimapped Read aligned to CLIP peak   -->
<!-- "Multimaped Reads %" - Percent of Multimapped reads that align to CLIP peak   -->
<!-- "Pvalue" - Sample specific peak P-Value compared to Control - MAnorm.   -->
<!-- "WT-Normalized count"- Normalized Peak count generated by MAnorm.   -->
<!-- "RoKO-Normalized count"- Control Sample normalized Peak count generated by MAnorm.   -->
<!-- "FC" - Log2 Fold Change (WT-RoKO) using MAnorm normalized counts.    -->
  
"Same Strand: Host_gene_name" - Name of Host Gene on same strand as CLIP reads  
"Same Strand: Host_gene_ensembl_id" - Ensemble Name of Host Gene on same strand as CLIP reads  
"Same Strand: RNA_type" - Transcript type of CLIP peak host gene on same strand as CLIP reads  
"Same Strand: Repeat_Type" - Type of repeat in Clip peak (determined by repeat masker) on same strand as CLIP reads  
"Same Strand: Intron Exon" - Gene feature of CLIP peak location in protein coding Gene on same strand as CLIP reads  
"Same Strand: RNA_type_Classification" - Summary of Peak Classification   
"Same Strand: ncRNA_Classification" - Summary of ncRNA Peak assignments  

"Opposite Strand: Host_gene_name" - Name of Host Gene on opposite strand as CLIP reads  
"Opposite Strand: Host_gene_ensembl_id" - Ensemble Name of Host Gene on opposite strand as CLIP reads  
"Opposite Strand: RNA_type" - Transcript type of CLIP peak host gene on opposite strand as CLIP reads  
"Opposite Strand: Repeat_Type" - Type of repeat in Clip peak (determined by repeat masker) on opposite strand as CLIP reads  
"Opposite Strand: Intron Exon" - Gene feature of CLIP peak location in protein coding Gene on opposite strand as CLIP reads  
"Opposite Strand: RNA_type_Classification" - Summary of Peak Classification  
"Opposite Strand: ncRNA_Classification" - Summary of ncRNA Peak assignments  

"Both Strand: RNA_type_Classification" - Summary of Peak Classification including both strand annotations  
      

<!-- "MM_number" - # of reads in CLIP peak that aligns to other locations Genome   -->
<!-- "MM_Alt_Peaktype" - # of Multimapped reads with Alternative Peak Classifications different from original Peak Classification   -->
<!-- "MM_Same_Peaktype" - # of Multimapped reads with the same Peak Classification as the original Peak Classification   -->
<!-- "MM_type" - Peak Classifications for alternative mappings of multimapped reads   -->
<!-- "MM_Prec_Alt_Peaktype" - % of Multimapped reads with Alternative Peak Classifications different from original Peak Classification   -->
<!-- "MM_Alt_Alignments" - Location alternate alignments for Multimapped reads   -->

<!-- "iCount_peak" - 1= peak is identified by iCount -->
<br>



<!-- ## Alignment Tracks   -->

<!-- `r toString(samplename)`_Clip_iCountcutadpt_all.unique.NH.mm.ddup.bam - MultiMapped + Uniquely aligned CLIP Reads   -->
<!-- `r toString(samplename)`_Clip_iCountcutadpt_all.unique.NH.mm.ddup.unique.bam - Uniquely aligned CLIP Reads   -->
<!-- `r toString(samplename)`_Clip_iCountcutadpt_all.unique.NH.mm.ddup.multimapped.bam - MultiMapped aligned CLIP Reads   -->

<br>
<br>

<!-- ## Annotation Tracks     -->

<!-- sftp://Biowulf.nih.gov/data/RBL_NCI/datashare/wolin/mESC_Clip/annotation_files/ -->

