---
title: ""
output: html_document
params:
  contrasts: '/Users/homanpj/OneDrive - National Institutes of Health/Loaner/Wolin/CLIP/CLIPpipeline/sam_test_master/contrastsx.txt'
  peak_in:  "/Users/homanpj/OneDrive - National Institutes of Health/Loaner/Wolin/CLIP/CLIPpipeline/sam_test_master/15_MAnorm/WT1vsKO1/WT1vsKO1_MAnormPeaksx.txt"
  PeakIdnt: "ALL"
  samplename: 'WT'
  background: 'KO' 
---
 
<!-- # Load libraries, set theme   -->

```{r envirment, include=F ,echo=F,warning=F,message=FALSE}
# rm(list=setdiff(ls(), "params"))
library(rmarkdown)
library(stringr)
library(rtracklayer)
library(VariantAnnotation,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(rmarkdown)
library(ggplot2,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library("viridis",quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(edgeR,quietly = T,verbose = F)
library('GenomicFeatures',quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library('rtracklayer',quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(matrixStats,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(plyr,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(tidyr,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(fitdistrplus,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(stringr,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(data.table)
library(reshape)
library(knitr)
library(stringi)
library(BSgenome)
library(biomaRt)
library(plotly)
library(tidyr)
library(GenomicRanges)
library(RColorBrewer)
library('gplots')
library(ggpubr)
library(circlize)
library('regioneR')
blank_theme <- theme_minimal()+
  theme(
  axis.title.x = element_blank(),
  axis.title.y = element_blank(),
  panel.border = element_blank(),
  panel.grid=element_blank(),
  axis.ticks = element_blank(),
  plot.title=element_text(size=14, face="bold")
  )
incd_rRNA=T  

```

<!-- # Set params, output   -->

```{r echo=F, message=FALSE, warning=FALSE, eval=T, include=F,results=F,}
contrasts=params$contrasts
peak_in=params$peak_in
compRow=params$compRow

# contrast=fread(contrasts, header=F, sep="\t",stringsAsFactors = F,data.table=F)
#   samplename=contrast[compRow,1]
#   background=contrast[compRow,2]
samplename=params$samplename
background=params$background
```

```{r echo=F, message=FALSE, warning=FALSE, eval=T, include=F,results=F,}

# 
# `Opposite Strand: RNA_type_Classification`="Oppo_Comb_type_exon"
# `Same Strand: RNA_type_Classification`="Same_Comb_type_exon"
# `Both Strand: RNA_type_Classification`="Comb_type_exon_Oppo"
# `Same Strand: Host_gene_name`="Same_external_gene_name"
# `Same Strand: ncRNA_Classification`="Same_Comb_type_ncRNA"

# Identification of CLIP peaks: `r toString(params$samplename)`    
Peaksdata2_anno = fread(peak_in, header=T, sep="\t",stringsAsFactors = F,data.table=F)
# unique(Peaksdata2_anno$`Same Strand: RNA_type_Classification`)
Peaksdata2_anno$`Same Strand: RNA_type_Classification` = 
    factor(Peaksdata2_anno$`Same Strand: RNA_type_Classification`, 
           levels = c("ncRNA", "protein_coding: exon", "Repeat Element","pseudogene",
                      "Antisense Feature","protein_coding: Intron","lncRNA","no Feature"))  
# unique(Peaksdata2_anno$`Opposite Strand: RNA_type_Classification`)
Peaksdata2_anno$`Opposite Strand: RNA_type_Classification` = 
    factor(Peaksdata2_anno$`Opposite Strand: RNA_type_Classification`, 
           levels = c("ncRNA", "protein_coding: exon", "Repeat Element","pseudogene",
                      "Antisense Feature","protein_coding: Intron","lncRNA","no Feature"))  
# unique(Peaksdata2_anno$`Both Strand: RNA_type_Classification`)
Peaksdata2_anno$`Both Strand: RNA_type_Classification` = 
    factor(Peaksdata2_anno$`Both Strand: RNA_type_Classification`, 
           levels = c("ncRNA", "protein_coding: exon", "Repeat Element","pseudogene",
                      "Antisense Feature","protein_coding: Intron","lncRNA","no Feature"))  
```
  
<!-- ## Processing data    -->
# Differential CLIP peaks identified by MAnorm (`r samplename`-`r background`)  

MAnorm used to identify Peaks preferentially expressed in the `r (samplename)` samples compared to `r (background)`.       

<br>  

## Identify preferentialy expressed CLIP peaks  


```{r ,fig.asp=.5, fig.align='center',fig.height=3.5, fig.width=7,echo=F,warning=F,eval=T,include=T}
nc_color=c("#1B9E77","cadetblue3",
           # "#2bdba6" ,
           "#D95F02", "#7570B3" ,"#E7298A", "#66A61E", "#E6AB02", "#A6761D", "#666666")
p=Peaksdata2_anno


p$`Both Strand: RNA_type_Classification`=factor(p$`Both Strand: RNA_type_Classification`, levels = c("ncRNA","rRNA", "protein_coding: exon", "Repeat Element","pseudogene","Antisense Feature","protein_coding: Intron","lncRNA","no Feature"))
p[(p$`Both Strand: RNA_type_Classification`=='ncRNA'&p$`Same Strand: ncRNA_Classification`=='rRNA'),'Both Strand: RNA_type_Classification']='rRNA'


g=ggplot(p,aes(x=FC,y=-log10(P_value),color=`Both Strand: RNA_type_Classification`,
                            text=paste0('ID: ',ID,'\nHostGene: ',`Same Strand: Host_gene_name`,
                                        '\n ',samplename,' normalized Reads: ', p[,paste0(samplename,'-normalized_Count')],
                                        '\n ',background,' normalized Reads: ', p[,paste0(background,'-normalized_Count')])
                            ))+geom_point(size=2)+
  theme_classic()+
  xlab(paste0("log<sub>2</sub> Fold Change (",samplename,"-",background,")"))+
  ylab(paste0("-log<sub>10</sub> P-Value"))+
  labs(color='RNA Type')+
  geom_hline(yintercept=-log10(.005),size=.5,col='black',linetype="dashed")+
  geom_vline(xintercept=(1),size=.5,col='black',linetype="dashed")+
  geom_vline(xintercept=(-1),size=.5,col='black',linetype="dashed")+
  # scale_color_brewer(palette = "Dark2") 
 scale_color_manual(values=nc_color)

ggplotly(g,tooltip = c("text"))



g=ggplot(p,aes(x=log2(p[,paste0(samplename,'-normalized_Count')]),y=FC,color=-log10(P_value),
                            text=paste0('ID: ',ID,
                                        '\nHostGene: ',`Same Strand: Host_gene_name`,
                                        '\n ',samplename,' normalized Reads: ', p[,paste0(samplename,'-normalized_Count')],
                                        '\n ',background,' normalized Reads: ', p[,paste0(background,'-normalized_Count')],
                                        '\n P-value: ',`P_value`)
                            ))+geom_point(size=2)+
  theme_classic()+
  xlab(paste0("log<sub>2</sub> normalized Counts (",samplename,")"))+
  ylab(paste0("log<sub>2</sub> Fold Change (",samplename,"-",background,")"))+
  # geom_hline(yintercept=-log10(.005),size=.5,col='black',linetype="dashed")+
  scale_colour_gradient(  low = ("blue"),  high = ("red"),
                           # midpoint = max(p$logPval)/2,
                           # midpoint = -log10(.05),
                           na.value = "grey50",guide = "colourbar")+
    geom_hline(yintercept=(1),size=.5,col='black',linetype="dashed")+
  geom_hline(yintercept=(-1),size=.5,col='black',linetype="dashed")
    
ggplotly(g,tooltip = c("text"))




g=ggplot(p,aes(x=log2(p[,paste0(samplename,'-normalized_Count')]),y=FC,color=`Both Strand: RNA_type_Classification`,
                            text=paste0('ID: ',ID,
                                        '\nHostGene: ',p[,'Same Strand: Host_gene_name'],
                                        '\n ',samplename,' normalized Reads: ', p[,paste0(samplename,'-normalized_Count')],
                                        '\n ',background,' normalized Reads: ', p[,paste0(background,'-normalized_Count')],
                                        '\n P-value: ',p[,'P_value'])
                            ))+geom_point(size=2)+
  theme_classic()+
  xlab(paste0("log<sub>2</sub> normalized Counts  (",samplename,")"))+
  ylab(paste0("log<sub>2</sub> Fold Change (",samplename,"-",background,")"))+
    labs(color='RNA Type')+
  # geom_hline(yintercept=-log10(.005),size=.5,col='black',linetype="dashed")+
  geom_hline(yintercept=(1),size=.5,col='black',linetype="dashed")+
  geom_hline(yintercept=(-1),size=.5,col='black',linetype="dashed")+
    # scale_color_brewer(palette = "Dark2") 
 scale_color_manual(values=nc_color) 
ggplotly(g,tooltip = c("text"))

```


```{r ,fig.asp=.5, fig.align='center',echo=F,warning=F,eval=T,include=T}

Peaksdata2_anno=Peaksdata2_anno[Peaksdata2_anno$P_value<.05&(Peaksdata2_anno$FC)>1,]
# SmplA_peaks_DEG=SmplA_Peaks[SmplA_Peaks$P_value<.05&SmplA_Peaks$FC>1,'ID']

```

<br>   
<br>  

## Peaks with a P-value < .05 and |log~2~(fold change)| > 1 were selected, resulting in `r toString(nrow(Peaksdata2_anno))` peaks.   

<br>  
<br>  

## Overview of targets by CLIP peak catagory   

```{r , echo=F, include=T, eval=T, fig.align='default', fig.ncol=2, fig.height=3.5, fig.width=7, out.height=200, out.width="50%", fig.show="hold", layout='l-screen', results = FALSE, message=FALSE , warning=F}
#Plot 1
if (params$PeakIdnt=='UNIQUE'){
  x1=ggplot(Peaksdata2_anno,aes(x=Counts_Unique,y=Length,color=`Both Strand: RNA_type_Classification`,text=`Same Strand: Host_gene_name`))+geom_point(size=1)+
    scale_x_continuous(trans = "log2",limits = c(.5,NA))+
    scale_y_continuous(trans = "log2",limits = c(NA,NA)) +
    theme_classic()+
    ylab("Peak Width")+xlab("Peak Counts (unique)")+ggtitle("Comparison of peak counts and Peak Width")+
    scale_color_brewer(palette = "Dark2") 
} else if (params$PeakIdnt=='ALL'){
  x1=ggplot(Peaksdata2_anno,aes(x=Counts_Multimappers_Scaled,y=Length,color=`Both Strand: RNA_type_Classification`,text=`Same Strand: Host_gene_name`))+geom_point(size=1)+
      scale_x_continuous(trans = "log2",limits = c(4,NA))+
      scale_y_continuous(trans = "log2",limits = c(NA,NA)) + 
    theme_classic()+
    ylab("Peak Width")+xlab("Peak Counts (Unique+ FracMM)")+ggtitle("Comparison of peak counts and Peak Width")+  labs(color='RNA Type')+
    scale_color_brewer(palette = "Dark2") 
}
#plot 2
x2=ggplot(Peaksdata2_anno,aes(x=Length))+geom_histogram(bins = 100)+
  scale_x_continuous(trans = "log2",limits = c(20,NA)) + 
  theme_classic()+ylab("# of peaks")+xlab("Peak Width")+ggtitle("Distribution of peak Width")
#plot 3
x3=ggplot(Peaksdata2_anno,aes(x=Length,line=`Both Strand: RNA_type_Classification`,color=`Both Strand: RNA_type_Classification`))+geom_density(size=.5)+
  scale_x_continuous(trans = "log2",limits = c(8,NA)) + 
  theme_classic()+ylab("Density")+xlab("Peak Width")+ggtitle("Distribution of peak Width")+  labs(color='RNA Type')+
  scale_color_brewer(palette = "Dark2") 
#plot 4
x4=ggplot(Peaksdata2_anno,aes(x=Length,line=`Both Strand: RNA_type_Classification`,color=`Both Strand: RNA_type_Classification`))+geom_freqpoly(bins = 100)+
  scale_x_continuous(trans = "log2",limits = c(8,NA)) + 
  theme_classic()+ylab("# of peaks")+xlab("Peak Width")+ggtitle("Distribution of peak Width")+  labs(color='RNA Type')+
  scale_color_brewer(palette = "Dark2") 
#plot 5
if (params$PeakIdnt=='UNIQUE'){
x5=ggplot(Peaksdata2_anno,aes(x=Counts_Unique,line=`Both Strand: RNA_type_Classification`,color=`Both Strand: RNA_type_Classification`))+geom_density()+
  scale_x_continuous(trans = "log2",limits = c(.25,NA)) + 
  theme_classic()+ylab("Density")+xlab("Peak Counts (unique)")+ggtitle("Distribution of peak counts")+ 
  labs(color='RNA Type')+
  scale_color_brewer(palette = "Dark2") 
} else if (params$PeakIdnt=='ALL'){
  x5=ggplot(Peaksdata2_anno,aes(x=Counts_Unique,line=`Both Strand: RNA_type_Classification`,color=`Both Strand: RNA_type_Classification`))+geom_density() +
    scale_x_continuous(trans = "log2",limits = c(.25,NA)) + 
    theme_classic()+ylab("Density")+xlab("Peak Counts (Unique+ FracMM)")+ggtitle("Distribution of peak counts") +
    labs(color='RNA Type') +
    scale_color_brewer(palette = "Dark2") 
}
#plot all graphs
x1;x2;x3;x4;x5
```

<br>  
<br>   

##  Number of peaks assigned to each catagory  


**Peak counts do include rRNA peaks**   

```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,eval=T,include=T}
p=Peaksdata2_anno
p$PlotColumn=p$`Both Strand: RNA_type_Classification`
if (incd_rRNA==FALSE) {
  p=p[!p$`Same Strand: ncRNA_Classification`%in%'rRNA',]
}
if (nrow(p)>0) {
#change types
p[is.na(p$PlotColumn),'PlotColumn']='no Feature'
colnames(p)[colnames(p)%in%'PlotColumn']='RNA_Type'
p$RNA=''
#plot
gg=ggplot(p,aes(fill=RNA_Type,x=RNA) ) +
  geom_bar(position="stack",width=.5) +
  theme_classic()+ggtitle("BioType - Counted by catagory\n(Each Clip Peak only Counted Once)") +
  scale_fill_brewer(palette = "Dark2") 
ggplotly(gg)
}else{print("No ncRNA")}

```


```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,eval=T,include=T}
p=Peaksdata2_anno
p$PlotColumn=p$`Both Strand: RNA_type_Classification`
###phil need else statement
#check if this makes sense
if (incd_rRNA==FALSE) {
  p=p[!p$`Same Strand: ncRNA_Classification`%in%'rRNA',]
} else{
  p=p
}
#filter
p=p[p$PlotColumn%in%'ncRNA','Same Strand: ncRNA_Classification',drop=F]
#if there are remaining types
if (nrow(p)>0) {
  p$RNA=""
  gg=ggplot(p,aes(fill=`Same Strand: ncRNA_Classification`,x=RNA) ) +
    geom_bar(width=.5)+theme_classic()+ggtitle("ncRNA only: # of peaks per catagory") +
    labs(fill='RNA Type') +
    scale_color_brewer(palette = "Paired")+scale_fill_brewer(palette = "Paired") + 
    blank_theme + 
    coord_polar("y",start=0)
  (gg) ###updating print to be consistent
  # ggplotly(gg)
}else{print("No ncRNA")}
 
```    

<br>  
<br> 

**Peak counts do not include rRNA peaks**   


 
```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,eval=T,include=T}
p=Peaksdata2_anno
p$PlotColumn=p$`Both Strand: RNA_type_Classification`
p=p[!p$`Same Strand: ncRNA_Classification`%in%'rRNA',]

if (nrow(p)>0) {
p[is.na(p$Comb_type_exon),'PlotColumn']='no Feature'
colnames(p)[colnames(p)%in%'PlotColumn']='RNA_Type'
p$RNA=''
#plot
gg=ggplot(p,aes(fill=RNA_Type,x=RNA) )+geom_bar(position="stack",width=.5) +
  theme_classic() +
  ggtitle("Number of peaks by catagory") +
  scale_fill_brewer(palette = "Dark2") 
ggplotly(gg)
}else{print("No ncRNA")}

```

```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,eval=T,include=T}
p=Peaksdata2_anno
p$PlotColumn=p$`Both Strand: RNA_type_Classification`
p=p[!p$`Same Strand: ncRNA_Classification`%in%'rRNA',]
p=p[p$PlotColumn%in%'ncRNA','Same Strand: ncRNA_Classification',drop=F]
#if there are remaining types
if (nrow(p)>0) {
  p$RNA=""
  gg=ggplot(p,aes(fill=`Same Strand: ncRNA_Classification`,x=RNA) ) +
    geom_bar(width=.5)+theme_classic()+
    ggtitle("Number of peaks by catagory : ncRNA only\n(Determined by unique reads only)") +
    labs(fill='ncRNA Type')+
    scale_color_brewer(palette = "Paired")+ 
    scale_fill_brewer(palette = "Paired") +  blank_theme+ 
    coord_polar("y",start=0)
  (gg)
  # ggplotly(gg)
}else{print("No ncRNA")}
```       

<br>  

**Peak counts do include rRNA peaks**   

```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,eval=T,include=T}
p=Peaksdata2_anno
p$PlotColumn=p$`Both Strand: RNA_type_Classification`
if (incd_rRNA==FALSE) {
  p=p[!p$`Same Strand: ncRNA_Classification`%in%'rRNA',]
}
if (nrow(p)>0) {
p[is.na(p$PlotColumn),'PlotColumn']='no Feature'
colnames(p)[colnames(p)%in%'PlotColumn']='RNA_Type'
p$RNA=''
gg=ggplot(p,aes(fill=RNA_Type,x=RNA) )+geom_bar(position="stack",width=.5)+theme_classic()+ggtitle("BioType - Counted by catagory\n(Each Clip Peak only Counted Once)")+
# theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15),legend.text = element_text( size=30))+
 scale_fill_brewer(palette = "Dark2") 
gg = ggplot(p,aes(fill=RNA_Type,x=RNA) ) +
  geom_bar(position="stack",width=.5) +
  theme_classic()+ggtitle("BioType - Counted by catagory\n(Each Clip Peak only Counted Once)")+
  scale_fill_brewer(palette = "Dark2") 
ggplotly(gg)
}else{print("No Features")}
```

<!-- <font size="1">Columns: RNA_type_Classification </font>      -->
<!-- <br/>      -->


```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,eval=T,include=T}
p=Peaksdata2_anno
p$PlotColumn=p$`Both Strand: RNA_type_Classification`
if (incd_rRNA==FALSE) {
  p=p[!p$`Same Strand: ncRNA_Classification`%in%'rRNA',]
}
p=p[p$PlotColumn%in%'ncRNA','Same Strand: ncRNA_Classification',drop=F]
if (nrow(p)>0) {
  
p$RNA=""
gg=ggplot(p,aes(fill=`Same Strand: ncRNA_Classification`,x=RNA) )+geom_bar(width=.5)+theme_classic()+ggtitle("ncRNA only: # of peaks per catagory")+  labs(fill='RNA Type')+
 scale_color_brewer(palette = "Paired")+scale_fill_brewer(palette = "Paired") +  blank_theme+ 
  coord_polar("y",start=0)
(gg)
}else(print('No ncRNA'))

 
``` 
<br>   
<br>   

## Number of reads assigned to each catagory   

#### Counts for RNA biotype  

<br>   

```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,include=T,eval=T}
nc_color=c("#1B9E77","#1fb487",
           "#2bdba6" ,
           "#D95F02", "#7570B3" ,"#E7298A", "#66A61E", "#E6AB02", "#A6761D", "#666666")
p=Peaksdata2_anno
p$PlotColumn=p$`Both Strand: RNA_type_Classification`
p$PlotColumn=as.character(p$PlotColumn)
p[p$`Same Strand: ncRNA_Classification`%in%'rRNA','PlotColumn']='rRNA'
p[p$`Same Strand: ncRNA_Classification`%in%'yRNA','PlotColumn']='yRNA'
p$PlotColumn=factor(p$PlotColumn, levels = c("ncRNA","rRNA","yRNA", "protein_coding: exon", "Repeat Element","pseudogene","Antisense Feature","protein_coding: Intron","lncRNA","no Feature"))
u=c("ncRNA","rRNA","yRNA", "protein_coding: exon", "Repeat Element","pseudogene","Antisense Feature","protein_coding: Intron","lncRNA","no Feature")
pcount=data.frame(u,1:length(u));colnames(pcount)=c('RNA_Type','Counts_All');pcount$Counts_Unique=NA;pcount$Counts_MM=NA;pcount$Peak_count=NA
for (x in 1:length(u)) {
  pam=p[p$PlotColumn%in%u[x],]
  pcount[x,1]=u[x]
  pcount[x,'Counts_Unique']=colSums(pam[,'Counts_Unique',drop=F],na.rm = T)
  pcount[x,'Counts_Multimappers_Scaled']=colSums(pam[,'Counts_Multimappers_Scaled',drop=F],na.rm = T)
    pcount[x,'Peak_count']=nrow(pam)
pcount$RNA=""
pcount$RNA_Type=factor(pcount$RNA_Type,
  levels = c("ncRNA","rRNA","yRNA",
              "protein_coding: exon", "Repeat Element",
              "pseudogene","Antisense Feature","protein_coding: Intron","lncRNA","no Feature"))
}
if (nrow(pcount)>0) {
gg = ggplot(pcount,aes(fill=RNA_Type,x=RNA,y=Counts_Unique)) +
  geom_bar(stat='identity',position="stack",width=.5)+theme_classic() +
  ggtitle("BioType - # of reads (unique) by catagory)") +
  scale_fill_manual(values=nc_color)  
ggplotly(gg)
#plot
gg = ggplot(pcount,aes(fill=RNA_Type,x=RNA,y=Counts_Multimappers_Scaled)) +
  geom_bar(stat='identity',position="stack",width=.5) +
  theme_classic() +
  ggtitle("BioType - # of reads (unique + Frac. MM) by catagory)")+
  scale_fill_manual(values=nc_color)
ggplotly(gg)
}else(print('No Features'))

```
<br>    

#### Counts for ncRNA      


```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,include=T,eval=T}
p=Peaksdata2_anno
p$PlotColumn=p$`Both Strand: RNA_type_Classification`
p$PlotColumn=as.character(p$PlotColumn)
p[p$`Same Strand: ncRNA_Classification`%in%'rRNA','PlotColumn']='rRNA'
p[p$`Same Strand: ncRNA_Classification`%in%'yRNA','PlotColumn']='yRNA'
p$PlotColumn=factor(p$PlotColumn, levels = c("ncRNA","rRNA","yRNA", "protein_coding: exon", "Repeat Element","pseudogene","Antisense Feature","protein_coding: Intron","lncRNA","no Feature"))
p=p[p$PlotColumn%in%'ncRNA',]
if (nrow(p)>0) {
  u=unique(p$`Same Strand: ncRNA_Classification`)
  pcount=data.frame(u,1:length(u));colnames(pcount)=c('Same_RNA_Subtype','Counts_All');pcount$Counts_Unique=NA;pcount$Counts_MM=NA;pcount$Peak_count=NA
  
  for (x in 1:length(u)) {
    pam=p[p$`Same Strand: ncRNA_Classification`%in%u[x],]
    pcount[x,1]=u[x]
    pcount[x,'Counts_Unique']=colSums(pam[,'Counts_Unique',drop=F],na.rm = T)
    pcount[x,'Counts_Multimappers_Scaled']=colSums(pam[,'Counts_Multimappers_Scaled',drop=F],na.rm = T)
    pcount[x,'Peak_count']=nrow(pam)
  }
if (nrow(pcount)>0) {
pcount$RNA=''
#plot
gg1 = ggplot(pcount,aes(fill=Same_RNA_Subtype,x=RNA,y=Counts_Unique) ) +
  geom_bar(stat='identity',width=.5)+theme_classic() +
  ggtitle("ncRNA only: # of reads (unique) by catagory") + 
  labs(color='RNA Type') +
  scale_color_brewer(palette = "Paired")+scale_fill_brewer(palette = "Paired") + 
  blank_theme + 
  coord_polar("y",start=0)
#plot
gg2 = ggplot(pcount,aes(fill=Same_RNA_Subtype,x=RNA,y=Counts_Multimappers_Scaled) ) +
  geom_bar(stat='identity',width=.5)+theme_classic() +
  ggtitle("ncRNA only: # of reads (unique + Frac. MM) by catagory") + 
  labs(color='RNA Type') +
  scale_color_brewer(palette = "Paired")+scale_fill_brewer(palette = "Paired") +
  blank_theme + 
  coord_polar("y",start=0)
plot(gg1)
plot(gg2)
}
}else(print('No ncRNA'))

```      
<br>   

<br>   

### Read Counts by peak for each catagory    
  
#### Unique Read counts (Count Unique reads only)   

<br>   

```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=8,echo=F,warning=F,eval=T, include=T}
p=Peaksdata2_anno
p$PlotColumn=p$`Both Strand: RNA_type_Classification`
###phil need else statement
if (incd_rRNA==FALSE) {
  p=p[!p$`Same Strand: ncRNA_Classification`%in%'rRNA',]
}
colnames(p)[colnames(p)%in%'PlotColumn']='RNA_Type'
if (nrow(p)>0) {
p$RNA=''
#plot
gg = ggplot(p,aes(x=RNA_Type,y=Counts_Unique,fill=RNA_Type)) +
    geom_violin()+theme_classic()+geom_jitter(height = 0, width = 0.1,size=0.5,aes(text=ID))+
  scale_y_continuous(trans = "log2") +
  ggtitle('Peak Counts (Unique) for\nby Feature') +
  ylab("Counts")+xlab("") +
      theme(plot.title = element_text(hjust = 0.5,size = 15),axis.title=element_text(size=15),axis.text=element_text(size=12))+theme(legend.position = "none") +
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = -45))
ggplotly(gg)
}else(print('No Features'))

```  
<br/>    

<br/>    

```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=8,echo=F,warning=F,eval=T, include=T}
p=Peaksdata2_anno
p$PlotColumn=p$`Both Strand: RNA_type_Classification`
p=p[p$PlotColumn%in%'ncRNA',]
if (nrow(p)>0) {
  p$ID=paste0(p$ID,"\n",p$Same_gene_name_comb)
  colnames(p)[colnames(p)%in%'Same Strand: ncRNA_Classification']='RNA_Type'
  #plot
  gg = ggplot(p,aes(x=RNA_Type,y=Counts_Unique,fill=RNA_Type)) +
    geom_violin()+theme_classic()+geom_jitter(height = 0, width = 0.1,size=0.5,aes(text=ID))+
    scale_y_continuous(trans = "log2") +
    ggtitle('Peak Counts (Unique)\nby Feature') +
    ylab("Counts") +
    xlab("")+
    theme(plot.title = element_text(hjust = 0.5,size = 15),axis.title=element_text(size=15),axis.text=element_text(size=12))+theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle = -45))+scale_fill_brewer(palette = "Paired")#+ylim(min(p$Counts_Unique),max(p$Counts_Unique))
  ggplotly(gg,tooltip = c("text", "y"))
}else(print('No ncRNA'))
```  
<br>   
<br>   

#### Fractional Multimap Read counts \n (Count Unique reads + (MultiMapped read / # of different mappings) )     

Does include rRNA
<br>   

```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=8,echo=F,warning=F,eval=T, include=T}
p=Peaksdata2_anno
p$PlotColumn=p$`Both Strand: RNA_type_Classification`
colnames(p)[colnames(p)%in%'PlotColumn']='RNA_Type'
if (nrow(p)>0) {
p$RNA=''
#plot
gg=ggplot(p,aes(x=RNA_Type,y=Counts_Multimappers_Scaled,fill=RNA_Type)) +
    geom_violin()+theme_classic()+geom_jitter(height = 0, width = 0.1,size=0.5,aes(text=ID))+
  scale_y_continuous(trans = "log2")+
  ggtitle('Peak Counts (Unique + Frac. MM)\nby Feature') +
  ylab("Counts")+xlab("")+
  theme(plot.title = element_text(hjust = 0.5,size = 15),axis.title=element_text(size=15),axis.text=element_text(size=12))+theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = -45))
ggplotly(gg)
}else {print("No Features")}

```  
<br/>    

<br/>    

```{r fig.asp=.5,fig.align='center',fig.height=6, fig.width=8,echo=F,warning=F,eval=T}
p=Peaksdata2_anno
p$PlotColumn=p$`Both Strand: RNA_type_Classification`
p=p[p$PlotColumn%in%'ncRNA',]
if (nrow(p)>0) {
  colnames(p)[colnames(p)%in%'Same Strand: ncRNA_Classification']='RNA_Type'
  p$ID=paste0(p$ID,"\n",p$Same_gene_name_comb)
  #plot
  gg=ggplot(p,aes(x=RNA_Type,y=Counts_Multimappers_Scaled,fill=RNA_Type) ) +
    geom_violin()+theme_classic()+geom_jitter(height = 0, width = 0.05,size=.5,aes(text=ID)) +
    scale_y_continuous(trans = "log2") +
    ggtitle('Peak Counts (Unique + Frac. MM)\nby ncRNA Type')+ylab("Counts") +
    xlab("") +
    theme(plot.title = element_text(hjust = 0.5,size = 15),axis.title=element_text(size=15),axis.text=element_text(size=12)) +
    theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle = -45))+scale_fill_brewer(palette = "Paired")#+ylim(min(p$Counts_Unique),max(p$Counts_Unique))
 ggplotly(gg,tooltip = c("text", "y"))
 
}else {print("No Features")}
```  
<br>   
<br>   

<br>  


## Opposite strand CLIP peak attributes   


```{r , fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F}
p=Peaksdata2_anno
p=p[p$`Both Strand: RNA_type_Classification`%in%'Antisense Feature',]
###phil need else statement
if (incd_rRNA==FALSE) {
  p=p[!p$`Same Strand: ncRNA_Classification`%in%'rRNA',]
}
if (nrow(p)>0) {
colnames(p)[colnames(p)%in%'Opposite Strand: RNA_type_Classification']='RNA_Type'
p$RNA=''
#plot
ggplot(p,aes(fill=RNA_Type,x=RNA_Type) )+geom_bar(stat="count")+theme_classic()+
  ggtitle("Opposite Strand Feature - Number of peaks by catagory\n(Determined by unique reads only)") +
  xlab("")+
  theme(plot.title = element_text(hjust = 0.5),axis.title=element_text(size=15),axis.text=element_text(size=10),legend.position = "none",axis.text.x = element_text(angle = -45,vjust = .5))+
  scale_fill_brewer(palette = "Dark2")
}else {print("No Features")}
```

```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T,incude=T}
p=Peaksdata2_anno
##phil need else statement
if (incd_rRNA==FALSE) {
  p=p[!p$`Same Strand: ncRNA_Classification`%in%'rRNA',]
}

p$transition=paste0(p$`Same Strand: RNA_type_Classification`,'->\n',p$`Opposite Strand: RNA_type_Classification`)
p=p[p$`Both Strand: RNA_type_Classification`%in%'Antisense Feature',]
if (nrow(p)>0) {
colnames(p)[colnames(p)%in%'transition']='RNA_Type'
p$Difference=''
#plot
gg = ggplot(p,aes(fill=`Same Strand: RNA_type_Classification`,x=factor(`Opposite Strand: RNA_type_Classification`)) ) +
  geom_bar(stat="count")+theme_classic()+
  ggtitle("Opposite Strand Feature Features: \n Compare Antisense and Snese stand Annotation")+xlab("Antisense strand Annotation")+ylab("Peak Count")+labs(fill = "Sense Stand Annotation")+
  theme(plot.title = element_text(hjust = 0.5,size=20),axis.title=element_text(size=15),axis.text=element_text(size=10),legend.position = "right",axis.text.x = element_text(angle = -45,vjust = .5))
ggplotly(gg)
}else {print("No Features")}
```



<br>  


# Output Data  

## Counts Table : 

"`r peak_in`"  


**Columns:**  
"ID" - CLIP peak location
"IDmerge" - CLIP regionns connected by spliced reads  
"strand" - CLIP peak stand  

"Counts_Unique" - Number of Unique reads aligned to CLIP peak  
"Counts_Multimappers_Scaled"  - number of reads with Unique reads =1 and Multimapped =1/#MM locations

"Pvalue" - `r samplename` specific peak P-Value compared to `r background` - MAnorm.
"`r samplename`-normalized_Count"- Normalized Peak count generated by MAnorm.  
"`r background`-normalized_Count"- Control Sample normalized Peak count generated by MAnorm.  
"FC" - Log2 Fold Change (`r samplename`,-`r background`) using MAnorm normalized counts.  

"Same Strand: Host_gene_name" - Name of Host Gene on same strand as CLIP reads  
"Same Strand: Host_gene_ensembl_id" - Ensemble Name of Host Gene on same strand as CLIP reads  
"Same Strand: RNA_type" - Transcript type of CLIP peak host gene on same strand as CLIP reads  
"Same Strand: Repeat_Type" - Type of repeat in Clip peak (determined by repeat masker) on same strand as CLIP reads  
"Same Strand: Intron Exon" - Gene feature of CLIP peak location in protein coding Gene on same strand as CLIP reads  
"Same Strand: RNA_type_Classification" - Summary of Peak Classification   
"Same Strand: ncRNA_Classification" - Summary of ncRNA Peak assignments  

"Opposite Strand: Host_gene_name" - Name of Host Gene on opposite strand as CLIP reads  
"Opposite Strand: Host_gene_ensembl_id" - Ensemble Name of Host Gene on opposite strand as CLIP reads  
"Opposite Strand: RNA_type" - Transcript type of CLIP peak host gene on opposite strand as CLIP reads  
"Opposite Strand: Repeat_Type" - Type of repeat in Clip peak (determined by repeat masker) on opposite strand as CLIP reads  
"Opposite Strand: Intron Exon" - Gene feature of CLIP peak location in protein coding Gene on opposite strand as CLIP reads  
"Opposite Strand: RNA_type_Classification" - Summary of Peak Classification  
"Opposite Strand: ncRNA_Classification" - Summary of ncRNA Peak assignments  

"Both Strand: RNA_type_Classification" - Summary of Peak Classification including both strand annotations 

<br>  
<br>  
<br>  
