---
title: ""
output: html_document
# params:
#   INpeaksAnno: "/Users/homanpj/OneDrive - National Institutes of Health/Loaner/Wolin/CLIP/HaCat_fCLIP2/"
#   INMAnorm: "/Users/homanpj/OneDrive - National Institutes of Health/Loaner/Wolin/CLIP/HaCat_fCLIP2/peaks_MAnorm/"
#   contrasts: '/Users/homanpj/OneDrive - National Institutes of Health/Loaner/Wolin/CLIP/HaCat_fCLIP2/contrasts.txt'
#   outDIR: "/Users/homanpj/OneDrive - National Institutes of Health/Loaner/Wolin/CLIP/HaCat_fCLIP2/peaks_MAnorm/"
#   ntmerge: "50nt"
#   readdepth: 5
#   Condense: FALSE
#   JoinJunc: TRUE
#   PeakIdnt: "all" #MultiMap, Unique, all
#   NameAdd: "Teststudio" ##"_SpliceAware_FullGenome_Transcme.SplicTrans_CollapseExons"
params:
  INpeaksAnno: 
    value: x
  INMAnorm:
    value: x
  contrasts:
    value: x
  OutDIR: 
    value: x
  ntmerge: 
    value: x
  readdepth:
    value: x
  species:
    value: x
  Condense: FALSE
  JoinJunc: TRUE
  PeakIdnt: FALSE
  NameAdd: FALSE
  sample_id:
    value: "WT"
  background_id:
    value: "KO"
  sample_pos:
    value: "WT_50nt_peakDepth5Test_Pos_MAvalues.xls"
  sample_neg:
    value: "KO_"
  peaks_mval_path:
    value: "WT_KO_manorm_values.txt"
  output_text:
    value: "WTvsKO_CLIP_MAnormSigPeaks_Peaks_50nt_peakDepth_MD15.txt"
  output_text2: 
    value: "WTvsKO_CLIP_MAnorm_Peaks_50nt_peakDepth_MD15.txt"
editor_options: 
  chunk_output_type: console
---

```{r echo=F, message=FALSE, warning=FALSE, include=F,eval=F}
###phil not sure what this file is?
#rmarkdown::render('fCLIP_WT_MD14_Pval.Rmd',
#                  encoding=encoding,
#                  output_file=paste0('fCLIP_',params$samplename,'_peakDepth',params$readdepth,params$NameAdd,'_MD14.html'))
```

 
```{r echo=F, message=FALSE, warning=FALSE, include=F, R.options=T}
###todo parse list as needed
library(rmarkdown)
library(stringr)
library(rtracklayer)
library(VariantAnnotation,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(rmarkdown)
library(ggplot2,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library("viridis",quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(edgeR,quietly = T,verbose = F)
library('GenomicFeatures',quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library('rtracklayer',quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(matrixStats,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(plyr,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(tidyr,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(fitdistrplus,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(stringr,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(data.table)
library(reshape)
library(knitr)
library(stringi)
library(BSgenome)
library(biomaRt)
library(plotly)
library(tidyr)
library(GenomicRanges)
library(RColorBrewer)
library('gplots')
library(ggpubr)
library(circlize)
library('regioneR')
library(scales)
library(dplyr)
```

```{r}
samplename=params$sample
background=params$background
  
INbam<-params$INbam 
INbamMM=params$INbamMM 
INpeaks=params$INpeaks 
inMAPQ=params$inMAPQ 
OUT=params$OUT 
ntmerge=params$ntmerge 
readdepth=as.numeric(params$readdepth )
Condense=as.logical(params$Condense)
NameAdd=params$NameAdd 
RunMAnorm=params$RunMAnorm


blank_theme <- theme_minimal()+
  theme(
  axis.title.x = element_blank(),
  axis.title.y = element_blank(),
  panel.border = element_blank(),
  panel.grid=element_blank(),
  axis.ticks = element_blank(),
  plot.title=element_text(size=14, face="bold")
  )

incd_rRNA=T


```

# Differencial CLIP peaks identified by MAnorm (`r samplename`-`r background`)  

MAnorm used to identify Peaks preferentially expressed in the `r (samplename)` samples compared to `r (background)`.       

<br>  
<br>  

## Identify preferentialy expressed CLIP peaks  

<!-- MAnorm calculates normalized counts and pvalues for peaks identified in WT sample    -->
###phil code moved to 09_manorm.R - keeping RMD's only for plotting

```{r ,fig.asp=.5, fig.align='center',echo=F,warning=F,eval=T,include=F}

# SmplA_Peaks=fread(paste0(params$INpeaksAnno,samplename,"_CLIP_",params$PeakIdnt,"Peaks",ntmerge,'_peakDepth',params$readdepth,params$NameAdd,"_MD15.txt"), header=T, sep="\t",stringsAsFactors = F,data.table=F)
#   SmplA_Peaks$`Both Strand: RNA_type_Classification`=factor(SmplA_Peaks$`Both Strand: RNA_type_Classification`, levels = c("ncRNA", "protein_coding: exon", "Repeat Element","pseudogene","lncRNA-exon","Antisense Feature","protein_coding: Intron","lncRNA-intron","lncRNA","no Feature"))
# 
# 
# SmplA_Peaks=separate(SmplA_Peaks,col = ID,into = c('chr','start'),sep = ':',remove = F)
# SmplA_Peaks=separate(SmplA_Peaks,col = start,into = c('start','end'),sep = '-',remove = F)
# SmplA_Peaks$start=as.numeric(SmplA_Peaks$start);SmplA_Peaks$end=as.numeric(SmplA_Peaks$end)
# SmplA_Peaks$Length=SmplA_Peaks$end-SmplA_Peaks$start
# 
# SmplA_Peaks.GR <- GRanges(seqnames = as.character(SmplA_Peaks$chr), ranges=IRanges(start = as.numeric(SmplA_Peaks$start), end = as.numeric(SmplA_Peaks$end)),strand = SmplA_Peaks$strand,ID=SmplA_Peaks$ID )
# 
# 
# SmplB_Peaks=fread(paste0(params$INpeaksAnno,background,"_CLIP_",params$PeakIdnt,"Peaks",ntmerge,'_peakDepth',params$readdepth,params$NameAdd,"_MD15.txt"), header=T, sep="\t",stringsAsFactors = F,data.table=F)
# 
# SmplB_Peaks=separate(SmplB_Peaks,col = ID,into = c('chr','start'),sep = ':',remove = F)
# SmplB_Peaks=separate(SmplB_Peaks,col = start,into = c('start','end'),sep = '-',remove = F)
# SmplB_Peaks$start=as.numeric(SmplB_Peaks$start);SmplB_Peaks$end=as.numeric(SmplB_Peaks$end)
# SmplB_Peaks$Length=SmplB_Peaks$end-SmplB_Peaks$start
# 
# SmplB_Peaks.GR <- GRanges(seqnames = as.character(SmplB_Peaks$chr), ranges=IRanges(start = as.numeric(SmplB_Peaks$start), end = as.numeric(SmplB_Peaks$end)),strand = SmplB_Peaks$strand,ID=SmplB_Peaks$ID )
# 
# 
# Neg_MAval=fread(paste0( NegMAnorm), header=T, sep="\t",stringsAsFactors = F,data.table=F)
#       colnames(Neg_MAval)=gsub("_Neg","",colnames(Neg_MAval))
#       Neg_MAval$strand="-"
# Pos_MAval=fread(paste0(PosMAnorm), header=T, sep="\t",stringsAsFactors = F,data.table=F)
#       colnames(Pos_MAval)=gsub("_Pos","",colnames(Pos_MAval))
#       Pos_MAval$strand="+"
# 
# MAval=rbind(Neg_MAval,Pos_MAval)
# MAval$start=MAval$start-1
# MAval$ID=paste0(MAval$chr,':',MAval$start,'-',MAval$end)
# MAval$Peak_Group=gsub("_Neg","",MAval$Peak_Group)
# MAval$Peak_Group=gsub("_Pos","",MAval$Peak_Group)
# colnames(MAval)=gsub(paste0('_',ntmerge,'_peakDepth',params$readdepth,params$NameAdd),"",colnames(MAval))
# colnames(MAval)=gsub(paste0('_',ntmerge,'_peakDepth',params$readdepth,'Test'),"",colnames(MAval))
# 
# SmplA_Peaks_MAval=merge(SmplA_Peaks,MAval,by='ID',suffixes=c("","_MAnorm"),all.x=T)
# 
# SmplA_Peaks_MAval[is.na(SmplA_Peaks_MAval$FC),'FC']=max(SmplA_Peaks_MAval$FC)
# SmplA_Peaks_MAval[is.na(SmplA_Peaks_MAval$P_value),'P_value']=1e-10
# 
# SmplA_Peaks_MAval$FC=log2(SmplA_Peaks_MAval[,paste0('normalized_read_density_in_',samplename)]+1)-log2(SmplA_Peaks_MAval[,paste0('normalized_read_density_in_',background)]+1)
# SmplA_Peaks_MAval$logPval=-log10(as.numeric(SmplA_Peaks_MAval$P_value))
# 
#   SmplA_Peaks_MAval$`Both Strand: RNA_type_Classification`=factor(SmplA_Peaks_MAval$`Both Strand: RNA_type_Classification`, levels = c("ncRNA", "protein_coding: exon", "Repeat Element","pseudogene","lncRNA-exon","Antisense Feature","protein_coding: Intron","lncRNA-intron","lncRNA","no Feature"))

```

```{r ,fig.asp=.5, fig.align='center',fig.height=3.5, fig.width=7,echo=F,warning=F,eval=T,include=T}
#read in peaks
SmplA_Peaks_MAval = peaks_mval_path

#graph Fold change : pvalue
###not sure if g_text will work like this - but help with readability
g_text = paste0('ID: ',ID,'\nHostGene: ',`Same Strand: Host_gene_name`,
                           '\n ',samplename,' normalized Reads: ',
                           SmplA_Peaks_MAval[,paste0('normalized_read_density_in_',samplename)],
                                        '\n ',background,' normalized Reads: ',
                           SmplA_Peaks_MAval[,paste0('normalized_read_density_in_',background)])
g = ggplot(SmplA_Peaks_MAval,
           aes(x=FC, 
               y=logPval, 
               color=`Both Strand: RNA_type_Classification`,
               text=g_text)) + 
  geom_point(size=2) +
  theme_classic()+
  xlab(paste0("log<sub>2</sub> Fold Change (",samplename,"-",background,")"))+
  ylab(paste0("-log<sub>10</sub> P-Value"))+
  labs(color='RNA Type')+
  geom_hline(yintercept=-log10(.005),size=.5,col='black',linetype="dashed")+
  geom_vline(xintercept=(1),size=.5,col='black',linetype="dashed")+
  geom_vline(xintercept=(-1),size=.5,col='black',linetype="dashed")+
  scale_color_brewer(palette = "Dark2")
ggplotly(g,tooltip = c("text"))

#graph normalized counts : fold change
g_text = paste0('ID: ',ID, 
                '\nHostGene: ',`Same Strand: Host_gene_name`,'\n ',
                samplename,' normalized Reads: ', 
                SmplA_Peaks_MAval[,paste0('normalized_read_density_in_',samplename)],'\n ',
                background,' normalized Reads: ', 
                SmplA_Peaks_MAval[,paste0('normalized_read_density_in_',background)], 
                '\n P-value: ',`P_value`)
g = ggplot(SmplA_Peaks_MAval,
         aes(x=log2(SmplA_Peaks_MAval[,paste0('normalized_read_density_in_',samplename)]),
             y=FC,
             color=logPval,
             text=g_text)) + 
  geom_point(size=2) +
  theme_classic() +
  xlab(paste0("log<sub>2</sub> normalized Counts (",samplename,")")) +
  ylab(paste0("log<sub>2</sub> Fold Change (",samplename,"-",background,")")) +
  scale_colour_gradient(  low = ("blue"),  high = ("red"),
                           na.value = "grey50",guide = "colourbar") +
  geom_hline(yintercept=(1),size=.5,col='black',linetype="dashed") +
  geom_hline(yintercept=(-1),size=.5,col='black',linetype="dashed")
ggplotly(g,tooltip = c("text"))

#graph normalized counts : fold change by RNA type
g_text = paste0('ID: ',ID, '\nHostGene: ', SmplA_Peaks_MAval[,'Same Strand: Host_gene_name'],'\n ',
                samplename,' normalized Reads: ', SmplA_Peaks_MAval[,paste0('normalized_read_density_in_',samplename)],'\n',
                background,' normalized Reads: ', SmplA_Peaks_MAval[,paste0('normalized_read_density_in_',background)],
                '\n P-value: ',SmplA_Peaks_MAval[,'P_value'])
g=ggplot(SmplA_Peaks_MAval,
         aes(x=log2(SmplA_Peaks_MAval[,paste0('normalized_read_density_in_',samplename)]),
             y=FC,color=`Both Strand: RNA_type_Classification`,
             text=g_text)) + 
  geom_point(size=2) +
  theme_classic() +
  xlab(paste0("log<sub>2</sub> normalized Counts  (",samplename,")")) +
  ylab(paste0("log<sub>2</sub> Fold Change (", samplename,"-",background,")")) +
  labs(color='RNA Type') +
  geom_hline(yintercept=(1),size=.5,col='black',linetype="dashed")+
  geom_hline(yintercept=(-1),size=.5,col='black',linetype="dashed")+
  scale_color_brewer(palette = "Dark2") 
ggplotly(g,tooltip = c("text"))

```

```{r ,fig.asp=.5, fig.align='center',echo=F,warning=F,eval=T,include=T}
#filter peaks
SmplA_Peaks_DEGtable = subset(SmplA_Peaks_MAval,P_value<.05 & FC>1)
```

<br>   
<br>  

**Peaks with a P-value < .05 and fold change > 1 were selected, resulting in `r toString(length(SmplA_Peaks_DEGtable))` peaks.**  

<br>   
<br>   

```{r Annotations, fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,echo=F,results=F}
###phil - if there are cols you want to remove, we should do that - it's dangerous to select from a col list like this
SmplA_Peaks_DEGtable2 = SmplA_Peaks_DEGtable[,c("ID","IDmerge","strand","chr","start","end" ,"Length",
                          'Counts_Unique', 'Counts_Multimappers_Scaled',
                          'P_value',paste0('normalized_read_density_in_',samplename),
                          paste0('normalized_read_density_in_',background),'FC',
                          "Same Strand: Host_gene_name","Same Strand: Host_gene_ensembl_id","Same Strand: RNA_type",
                          'Same Strand: Repeat_Type',"Same Strand: Intron Exon",'Same Strand: RNA_type_Classification',
                          'Same Strand: ncRNA_Classification',"Opposite Strand: Host_gene_name",
                          "Opposite Strand: Host_gene_ensembl_id","Opposite Strand: RNA_type",'Opposite Strand: Repeat_Type',
                          "Opposite Strand: Intron Exon",'Opposite Strand: RNA_type_Classification',
                          'Opposite Strand: ncRNA_Classification','Both Strand: RNA_type_Classification')]

###phil same for this rename; if it's necessary we can do a select (new_name = old_name) to match cols more concretely
colnames(SmplA_Peaks_DEGtable2)=c("ID","IDmerge","strand","chr","start","end","Length",
                                'Counts_Unique','Counts_fracMM',
                                'P_value',paste0('normalized_read_density_in_',samplename),
                                paste0('normalized_read_density_in_',background),'FC',
                                "Same_gene_name_comb","Same_ensembl_gene_id",
                                "Same_type_comb","Same_Repeat_comb",'Same_feature',
                                "Same_Comb_type_exon","Same_Comb_type_ncRNA",
                                "Oppo_gene_name_comb","Oppo_ensembl_gene_id",
                                "Oppo_type_comb","Oppo_Repeat_comb",'Oppo_feature',
                                "Oppo_Comb_type_exon","Oppo_Comb_type_ncRNA","Comb_type_exon_Oppo")

###fix variable swap
Peaksdata2_anno=SmplA_Peaks_DEGtable2

###phil do we need these as factors?
Peaksdata2_anno$Same_Comb_type_exon = factor(Peaksdata2_anno$Same_Comb_type_exon, 
                                             levels = c("ncRNA","rRNA","yRNA", "protein_coding: exon", "Repeat Element",
                                                        "pseudogene","Antisense Feature","protein_coding: Intron",
                                                        "lncRNA","no Feature"))
Peaksdata2_anno$Comb_type_exon_Oppo = factor(Peaksdata2_anno$Comb_type_exon_Oppo, 
                                             levels = c("ncRNA", "protein_coding: exon", 
                                                        "Repeat Element","pseudogene","lncRNA-exon",
                                                        "Antisense Feature","protein_coding: Intron",
                                                        "lncRNA-intron","lncRNA","no Feature"))
```


<!-- ### Asigning Clip peak attributes    -->

<!-- Not all Peaks overlap with a single feature so peak assignments were assigned by priority:   -->

<!-- **ncRNA > Protein coding : Exonic > repeats > Pseudogene > Antisense Feature > Protein Coding : Intronic > lncRNA > no Feature**    -->

<!-- **ncRNA > protein_coding: Exonic > Repeat Element > pseudogene > lncRNA: Exonic > Antisense Feature > protein_coding: Intronic > lncRNA: Intronic > no Feature** -->

<!-- rDNA transcript (BK000964.3) was included in reference genome but not included in peak counts. rDNA will affect multimapping count and will show up in alternative mapping location.   	 -->

<!-- All annotations from RNA type, Repeat regions, and Intronic/exonic regions are annotated in the Table.    -->

<!-- <br>   -->


<!-- **A Large number of Exonic CLIP peaks are from same Gene so Exonic CLIP peaks in same gene were combined into one 'Peak' resulting in `r toString(nrow(Peaksdata2_anno))` peaks.** -->


## Overview of targets by CLIP peak catagory   


```{r , echo=F, include=T, eval=T, fig.align='default', fig.ncol=2, fig.height=3.5, fig.width=7, out.height=200, out.width="50%", fig.show="hold", layout='l-screen', results = FALSE, message=FALSE , warning=F}
###phil we'll need to update these based on using unique/all - not sure if that means the 
#graphs should change too...
#graph peak counts
if (params$PeakIdnt=='Unique'){
  x1 = ggplot(Peaksdata2_anno,
              aes(x=Counts_Unique,
                  y=Length,
                  color=Comb_type_exon_Oppo,
                  text=Same_gene_name_comb))+
    geom_point(size=1)+
    scale_x_continuous(trans = "log2")+
    scale_y_continuous(trans = "log2") +
    theme_classic()+
    ylab("Peak Width")+
    xlab("Peak Counts (unique)")+
    ggtitle("Comparison of peak counts and Peak Width")+
    labs(color='RNA Type')+
    scale_color_brewer(palette = "Dark2") 
} else if (params$PeakIdnt=='MultiMap'){
  x1 = ggplot(Peaksdata2_anno,
              aes(x=Counts_MM,
                  y=Length,
                  color=Comb_type_exon_Oppo,
                  text=Same_gene_name_comb))+
    geom_point(size=1)+
    scale_x_continuous(trans = "log2")+
    scale_y_continuous(trans = "log2") + 
    theme_classic()+
    ylab("Peak Width")+
    xlab("Peak Counts (MultiMap)")+
    ggtitle("Comparison of peak counts and Peak Width")+
    labs(color='RNA Type')+
    scale_color_brewer(palette = "Dark2") 
}else if (params$PeakIdnt=='all'){
  x1 = ggplot(Peaksdata2_anno,
              aes(x=Counts_fracMM,
                  y=Length,
                  color=Comb_type_exon_Oppo,
                  text=Same_gene_name_comb))+
    geom_point(size=1)+
    scale_x_continuous(trans = "log2")+
    scale_y_continuous(trans = "log2") + 
    theme_classic()+
    ylab("Peak Width")+
    xlab("Peak Counts (Unique+ FracMM)")+
    ggtitle("Comparison of peak counts and Peak Width")+
    labs(color='RNA Type')+
    scale_color_brewer(palette = "Dark2") 
}

#graph peak width
x2 = ggplot(Peaksdata2_anno,
            aes(x=Length))+
  geom_histogram(bins = 100)+
  scale_x_continuous(trans = "log2") + 
  theme_classic()+ylab("# of peaks")+
  xlab("Peak Width")+
  ggtitle("Distribution of peak Width")

#plot peak width log transformed
x3 = ggplot(Peaksdata2_anno,
            aes(x=Length,
                line=Comb_type_exon_Oppo,
                color=Comb_type_exon_Oppo))+
  geom_density(size=.5)+
  scale_x_continuous(trans = "log2") + 
  theme_classic()+ylab("Density")+
  xlab("Peak Width")+
  ggtitle("Distribution of peak Width")+
  labs(color='RNA Type')+
  scale_color_brewer(palette = "Dark2") 

#plot peak width log transformed by RNA type
x4 = ggplot(Peaksdata2_anno,
            aes(x=Length,
                line=Comb_type_exon_Oppo,
                color=Comb_type_exon_Oppo))+
  geom_freqpoly(bins = 100)+
  scale_x_continuous(trans = "log2") + 
  theme_classic()+ylab("# of peaks")+
  xlab("Peak Width")+
  ggtitle("Distribution of peak Width")+
  labs(color='RNA Type')+
  scale_color_brewer(palette = "Dark2") 

###phil same here unique/mm/all
if (params$PeakIdnt=='Unique'){
  x5=ggplot(Peaksdata2_anno,
            aes(x=Counts_Unique,
                line=Comb_type_exon_Oppo,
                color=Comb_type_exon_Oppo))+
    geom_density()+
    scale_x_continuous(trans = "log2") + 
    theme_classic()+ylab("Density")+
    xlab("Peak Counts (unique)")+
    ggtitle("Distribution of peak counts")+
    labs(color='RNA Type')+
    scale_color_brewer(palette = "Dark2") 
} else if (params$PeakIdnt=='MultiMap'){
  x5 = ggplot(Peaksdata2_anno,
              aes(x=Counts_Unique,
                  line=Comb_type_exon_Oppo,
                  color=Comb_type_exon_Oppo))+
    geom_density()+scale_x_continuous(trans = "log2") + 
    theme_classic()+ylab("Density")+
    xlab("Peak Counts (Counts_MM)")+
    ggtitle("Distribution of peak counts")+
    labs(color='RNA Type')+
    scale_color_brewer(palette = "Dark2") 
} else if (params$PeakIdnt=='all'){
  x5 = ggplot(Peaksdata2_anno,
              aes(x=Counts_Unique,
                  line=Comb_type_exon_Oppo,
                  color=Comb_type_exon_Oppo))+
    geom_density()+scale_x_continuous(trans = "log2") + 
    theme_classic()+ylab("Density")+
    xlab("Peak Counts (Unique+ FracMM)")+
    ggtitle("Distribution of peak counts")+ 
    labs(color='RNA Type')+
    scale_color_brewer(palette = "Dark2") 
}

#print all graphs
x1;x2;x3;x4;x5
```

<br>  
<br>   

<!-- ##  Peak Attributes    -->

##  Number of peaks assigned to each catagory  
<br>   

**Peak counts do not include rRNA peaks**   

###phil - this whole section is very similar to what we're doing in MD15_BW
#if its necessary to repeat, we may want to think about pulling this code into an R
#script, and then pulling the images back into the RMD files. it'll save us the comp
#of having to run the same code twice
```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,eval=T,include=T}

###fix variable swap
p=Peaksdata2_anno

###phil we are doing the same thing as with MD15_BW - not sure why we are swapping cols?
p$Same_Comb_type_exon=p$Comb_type_exon_Oppo
p=subset(p,!(Same_Comb_type_ncRNA == 'rRNA'))
p[is.na(p$Comb_type_exon),'Same_Comb_type_exon']='no Feature'
colnames(p)[colnames(p)%in%'Same_Comb_type_exon']='RNA_Type'
###phil same q - why are we clearing the col
p$RNA=''

#plot number of unique peaks
gg = ggplot(p,
            aes(fill=RNA_Type
                ,x=RNA))+
  geom_bar(position="stack",width=.5)+
  theme_classic()+ggtitle("Number of peaks by catagory\n(Determined by unique reads only)")+
  scale_fill_brewer(palette = "Dark2") 
ggplotly(gg)
```


<!-- <font size="1">Columns: RNA_type_Classification </font>      -->
<!-- <br/>      -->


```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,eval=T,include=T}

###fix variable swap
p=Peaksdata2_anno
p$Same_Comb_type_exon=p$Comb_type_exon_Oppo
p=subset(p,!(Same_Comb_type_ncRNA == 'rRNA')) ###todo fix these later with %>%
p=p[p$Same_Comb_type_exon%in%'ncRNA','Same_Comb_type_ncRNA',drop=F]
p$RNA=""

###phil - need an else here
if (nrow(p)>0) {
  gg=ggplot(p,
            aes(fill=Same_Comb_type_ncRNA,x=RNA) )+
    geom_bar(width=.5)+
    theme_classic()+
    ggtitle("Number of peaks by catagory : ncRNA only\n(Determined by unique reads only)")+  labs(fill='ncRNA Type')+
    scale_color_brewer(palette = "Paired")+ 
    scale_fill_brewer(palette = "Paired") +  blank_theme+ 
    coord_polar("y",start=0)
  gg
}

```       

<br>  
<br>   
<!-- ## Overview of Ro targets by CLIP peak attributes (no rRNA peaks) -->

**Peak counts do include rRNA peaks**   

```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,eval=T,include=T}
p=Peaksdata2_anno
p$Same_Comb_type_exon=p$Comb_type_exon_Oppo

if (incd_rRNA==F) {
  p=p[!p$Same_Comb_type_ncRNA%in%'rRNA',]
}

p[is.na(p$Same_Comb_type_exon),'Same_Comb_type_exon']='no Feature'
colnames(p)[colnames(p)%in%'Same_Comb_type_exon']='RNA_Type'
p$RNA=''
gg=ggplot(p,aes(fill=RNA_Type,x=RNA) )+geom_bar(position="stack",width=.5)+theme_classic()+ggtitle("BioType - Counted by catagory\n(Each Clip Peak only Counted Once)")+
 scale_fill_brewer(palette = "Dark2") 
ggplotly(gg)

```

<!-- <font size="1">Columns: RNA_type_Classification </font>      -->
<!-- <br/>      -->

```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,eval=T,include=T}

p=Peaksdata2_anno
p$Same_Comb_type_exon=p$Comb_type_exon_Oppo
if (incd_rRNA==F) {
  p=p[!p$Same_Comb_type_ncRNA%in%'rRNA',]
}

p=p[p$Same_Comb_type_exon%in%'ncRNA','Same_Comb_type_ncRNA',drop=F]
if (nrow(p)>0) {
  
p$RNA=""
gg=ggplot(p,aes(fill=Same_Comb_type_ncRNA,x=RNA) )+geom_bar(width=.5)+theme_classic()+ggtitle("ncRNA only: # of peaks per catagory")+  labs(fill='ncRNA Type')+
 scale_color_brewer(palette = "Paired")+scale_fill_brewer(palette = "Paired") +  blank_theme+ 
  coord_polar("y",start=0)
(gg)

}

 
```    


<!-- ## Calculate average MAPQ for each peak    -->

<!-- We can plot the avarage Mapping Quality (MAPQ) of each peak to identify the confidence of read mapping. We see that repeat regions have generally lower mapping quality.     -->

<!-- <br>   -->
```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=F,incude=T}

p=Peaksdata2_anno


colnames(p)[colnames(p)%in%'Comb_type_exon_Oppo']='RNA_Type'
p$RNA=''
 gg=ggplot(p,aes(x=RNA_Type,y=Avg_mapq,fill=RNA_Type) )+geom_violin()+theme_classic()+geom_jitter(height = 0, width = 0.1,size=1)+
  ggtitle('Average MAPQ for Ro-Clip peak\nby Feature')+ylab("Avg_mapq - Ro-Clip")+xlab("")+
theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15))+theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle = -45))
gg


p=Peaksdata2_anno
p=p[p$Comb_type_exon_Oppo%in%'Repeat Element',]

p=p[,c('ID','Avg_mapq',colnames(p)[grep('Same_Repeat_',colnames(p))])]
p2=p[,colnames(p)[grep('Same_Repeat_',colnames(p))]]
p2[,colnames(p2)[grep('Same_Repeat_',colnames(p2))]]=as.numeric(is.na(p2[,colnames(p2)[grep('Same_Repeat_',colnames(p2))]])==F)

p2[p2[,colnames(p2)[grep('Same_Repeat_',colnames(p2))]]==0]=NA
p2=(p2*p$Avg_mapq)

p2=melt.data.frame(p2,id.vars = NULL)
p2$variable=gsub("Same_Repeat_","",p2$variable)


 gg=ggplot(p2,aes(x=variable,y=value,fill=variable) )+geom_violin()+theme_classic()+geom_jitter( width = 0.1,size=1)+
  ggtitle('Same Strand: Repeat Element\nAverage MAPQ for Ro-Clip peak')+ylab("Avg_mapq - Ro-Clip")+xlab("")+
theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15))+theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle = -45))+scale_fill_brewer(palette = "Paired")

gg
 
 
p=Peaksdata2_anno
p=p[p$Comb_type_exon_Oppo%in%'Repeat Element',]

p=p[,c('ID','Avg_mapq',colnames(p)[grep('Oppo_Repeat_',colnames(p))])]
p2=p[,colnames(p)[grep('Oppo_Repeat_',colnames(p))]]
p2[,colnames(p2)[grep('Oppo_Repeat_',colnames(p2))]]=as.numeric(is.na(p2[,colnames(p2)[grep('Oppo_Repeat_',colnames(p2))]])==F)

p2[p2[,colnames(p2)[grep('Oppo_Repeat_',colnames(p2))]]==0]=NA
p2=(p2*p$Avg_mapq)

p2=melt(p2,id.vars = NULL)
p2$variable=gsub("Oppo_Repeat_","",p2$variable)

gg= ggplot(p2,aes(x=variable,y=value,fill=variable) )+geom_violin()+theme_classic()+geom_jitter( width = 0.1,size=1)+
  ggtitle('Opposite strand feature: Repeat Element\nAverage MAPQ for Ro-Clip peak')+ylab("Avg_mapq - Ro-Clip")+xlab("")+
theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15))+theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle = -45))
 
gg

```

<!-- <br>    -->

<!-- For comparison the average peak MAPQ is ploted for each ncRNA subtype.    -->

<!-- <br>   -->

```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=F,incude=T}

 
 p=Peaksdata2_anno

p=p[p$Comb_type_exon_Oppo%in%'ncRNA',]
if (nrow(p)>0) {
  
colnames(p)[colnames(p)%in%'Same_Comb_type_ncRNA']='RNA_Type'


 gg=ggplot(p,aes(x=RNA_Type,y=Avg_mapq,fill=RNA_Type) )+geom_violin()+theme_classic()+geom_jitter( width = 0.1,size=1)+
  ggtitle('Average MAPQ for Ro-Clip peak\nby ncRNA Type')+ylab("Average MAPQ - Ro-Clip")+xlab("")+
theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15))+theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle = -45))+scale_fill_brewer(palette = "Paired")

gg
}
 
 
```
 
 
<!-- <br>    -->
<!-- <br>    -->

<br>  
<br>   
<br>   

## Number of reads assigned to each catagory   

#### Counts for RNA biotype  

<br>   

```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,include=T,eval=T}

nc_color=c("#1B9E77","#1fb487",
           "#2bdba6" ,
           "#D95F02", "#7570B3" ,"#E7298A", "#66A61E", "#E6AB02", "#A6761D", "#666666")

p=Peaksdata2_anno
p$Same_Comb_type_exon=p$Comb_type_exon_Oppo

p$Same_Comb_type_exon=as.character(p$Same_Comb_type_exon)
p[p$Same_Comb_type_ncRNA%in%'rRNA','Same_Comb_type_exon']='rRNA'
p[p$Same_Comb_type_ncRNA%in%'yRNA','Same_Comb_type_exon']='yRNA'

p$Same_Comb_type_exon=factor(p$Same_Comb_type_exon, levels = c("ncRNA","rRNA","yRNA", "protein_coding: exon", "Repeat Element","pseudogene","Antisense Feature","protein_coding: Intron","lncRNA","no Feature"))


u=unique(p$Same_Comb_type_exon)
pcount=data.frame(u,1:length(u));colnames(pcount)=c('RNA_Type','Counts_All');pcount$Counts_Unique=NA;pcount$Counts_MM=NA;pcount$Peak_count=NA
for (x in 1:length(u)) {
  pam=p[p$Same_Comb_type_exon%in%u[x],]
  pcount[x,1]=u[x]
  pcount[x,'Counts_Unique']=colSums(pam[,'Counts_Unique',drop=F],na.rm = T)
   pcount[x,'Counts_fracMM']=colSums(pam[,'Counts_fracMM',drop=F],na.rm = T)
    pcount[x,'Peak_count']=nrow(pam)

}
pcount$RNA=""



if (params$PeakIdnt=='MultiMap'){

gg=
ggplot(pcount,aes(fill=RNA_Type,x=RNA,y=Counts_MM) )+geom_bar(stat='identity',position="stack",width=.5)+theme_classic()+ggtitle("BioType - # of reads (MultiMap) by catagory)")+
 scale_fill_manual(values=nc_color)

ggplotly(gg)
}


gg=
ggplot(pcount,aes(fill=RNA_Type,x=RNA,y=Counts_Unique) )+geom_bar(stat='identity',position="stack",width=.5)+theme_classic()+ggtitle("BioType - # of reads (unique) by catagory)")+
 scale_fill_manual(values=nc_color)

ggplotly(gg)
# }



gg=
ggplot(pcount,aes(fill=RNA_Type,x=RNA,y=Counts_fracMM) )+geom_bar(stat='identity',position="stack",width=.5)+theme_classic()+ggtitle("BioType - # of reads (unique + Frac. MM) by catagory)")+
scale_fill_manual(values=nc_color)
ggplotly(gg)
```
<br>    

#### Counts for ncRNA      


```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,include=T,eval=T}

p=Peaksdata2_anno
p$Same_Comb_type_exon=p$Comb_type_exon_Oppo

p$Same_Comb_type_exon=as.character(p$Same_Comb_type_exon)
p[p$Same_Comb_type_ncRNA%in%'rRNA','Same_Comb_type_exon']='rRNA'
p[p$Same_Comb_type_ncRNA%in%'yRNA','Same_Comb_type_exon']='yRNA'

p$Same_Comb_type_exon=factor(p$Same_Comb_type_exon, levels = c("ncRNA","rRNA","yRNA", "protein_coding: exon", "Repeat Element","pseudogene","Antisense Feature","protein_coding: Intron","lncRNA","no Feature"))

p=p[p$Same_Comb_type_exon%in%'ncRNA',]
if (nrow(p)>0) {
  
u=unique(p$Same_Comb_type_ncRNA)
pcount=data.frame(u,1:length(u));colnames(pcount)=c('Same_RNA_Subtype','Counts_All');pcount$Counts_Unique=NA;pcount$Counts_MM=NA;pcount$Peak_count=NA
for (x in 1:length(u)) {
  pam=p[p$Same_Comb_type_ncRNA%in%u[x],]
  pcount[x,1]=u[x]
  pcount[x,'Counts_Unique']=colSums(pam[,'Counts_Unique',drop=F],na.rm = T)
   pcount[x,'Counts_fracMM']=colSums(pam[,'Counts_fracMM',drop=F],na.rm = T)
  pcount[x,'Peak_count']=nrow(pam)
}

pcount$RNA=''


gg1=ggplot(pcount,aes(fill=Same_RNA_Subtype,x=RNA,y=Counts_Unique) )+geom_bar(stat='identity',width=.5)+theme_classic()+ggtitle("ncRNA only: # of reads (unique) by catagory")+  labs(fill='ncRNA Type')+
 scale_color_brewer(palette = "Paired")+scale_fill_brewer(palette = "Paired") +  blank_theme+ 
  coord_polar("y",start=0)


gg2=ggplot(pcount,aes(fill=Same_RNA_Subtype,x=RNA,y=Counts_fracMM) )+geom_bar(stat='identity',width=.5)+theme_classic()+ggtitle("ncRNA only: # of reads (unique + Frac. MM) by catagory")+  labs(fill='RNA Type')+
 scale_color_brewer(palette = "Paired")+scale_fill_brewer(palette = "Paired") +  blank_theme+ 
  coord_polar("y",start=0)
plot(gg1)
plot(gg2)
}
```      
<!-- <font size="1">Columns: RNA_type_Classification + Counts_Unique </font>      -->

<!-- <br>    -->

<!-- ### Number of reads for each catagory   -->

<br>   

```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,eval=T,include=F}

p=Peaksdata2_anno
if (incd_rRNA==F) {
  p=p[!p$Oppo_Comb_type_ncRNA%in%'rRNA',]
}

u=unique(p$Oppo_Comb_type_exon)
pcount=data.frame(u,1:length(u));colnames(pcount)=c('RNA_Type','Counts_All');pcount$Counts_Unique=NA;pcount$Counts_MM=NA;pcount$Peak_count=NA
for (x in 1:length(u)) {
  pam=p[p$Oppo_Comb_type_exon%in%u[x],]
  pcount[x,1]=u[x]
  pcount[x,'Counts_Unique']=colSums(pam[,'Counts_Unique',drop=F],na.rm = T)
  pcount[x,'Counts_fracMM']=colSums(pam[,'Counts_fracMM',drop=F],na.rm = T)
    pcount[x,'Peak_count']=nrow(pam)

}
pcount$RNA=""

gg=
ggplot(pcount,aes(fill=RNA_Type,x=RNA,y=Counts_Unique) )+geom_bar(stat='identity',position="stack",width=.5)+theme_classic()+ggtitle("BioType - # of reads (unique) by catagory)")+
 scale_fill_brewer(palette = "Dark2") 
ggplotly(gg)

gg=
ggplot(pcount,aes(fill=RNA_Type,x=RNA,y=Counts_fracMM) )+geom_bar(stat='identity',position="stack",width=.5)+theme_classic()+ggtitle("BioType - # of reads (Unique + Scaled Multimap) by catagory)")+
 scale_fill_brewer(palette = "Dark2")
ggplotly(gg)
```


```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,eval=T,include=F}

p=Peaksdata2_anno
if (incd_rRNA==F) {
  p=p[!p$Oppo_Comb_type_ncRNA%in%'rRNA',]
}
p=p[p$Oppo_Comb_type_exon%in%'ncRNA',]
if (nrow(p)>0) {

u=unique(p$Oppo_Comb_type_ncRNA)
pcount=data.frame(u,1:length(u));colnames(pcount)=c('RNA_Subtype','Counts_All');pcount$Counts_Unique=NA;pcount$Counts_MM=NA;pcount$Peak_count=NA
for (x in 1:length(u)) {
  pam=p[p$Oppo_Comb_type_ncRNA%in%u[x],]
  pcount[x,1]=u[x]
  pcount[x,'Counts_Unique']=colSums(pam[,'Counts_Unique',drop=F],na.rm = T)
   pcount[x,'Counts_fracMM']=colSums(pam[,'Counts_fracMM',drop=F],na.rm = T)
  pcount[x,'Peak_count']=nrow(pam)
}


pcount$RNA=''


gg=ggplot(pcount,aes(fill=RNA_Subtype,x=RNA,y=Counts_Unique) )+geom_bar(stat='identity',width=.5)+theme_classic()+ggtitle("ncRNA only: # of reads (unique) by catagory")+
 scale_color_brewer(palette = "Paired")+scale_fill_brewer(palette = "Paired") +  blank_theme+ 
  coord_polar("y",start=0)
(gg)


gg=ggplot(pcount,aes(fill=RNA_Subtype,x=RNA,y=Counts_fracMM) )+geom_bar(stat='identity',width=.5)+theme_classic()+ggtitle("ncRNA only: # of reads (Scaled Multimap) by catagory")+
 scale_color_brewer(palette = "Paired")+scale_fill_brewer(palette = "Paired") +  blank_theme+ 
  coord_polar("y",start=0)
(gg)

}
```      
<!-- <font size="1">Columns: Opposite Strand: RNA_type </font>      -->

### Read Counts by peak for each catagory    
  
#### Unique Read counts (Count Unique reads only)   

<br>   

```{r fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T, include=T}
p=Peaksdata2_anno
p$Same_Comb_type_exon=p$Comb_type_exon_Oppo
if (incd_rRNA==F) {
  p=p[!p$Same_Comb_type_ncRNA%in%'rRNA',]
}

colnames(p)[colnames(p)%in%'Same_Comb_type_exon']='RNA_Type'
p$RNA=''
 gg=ggplot(p,aes(x=RNA_Type,y=Counts_Unique,fill=RNA_Type) )+geom_violin()+theme_classic()+geom_jitter(height = 0, width = 0.1,size=1)+
  scale_y_continuous(trans = "log2")+
  ggtitle('Peak Counts (Unique) for Ro-Clip\nby Feature')+ylab("Counts - Ro-Clip")+xlab("")+
theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15))+theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle = -45))
# 
gg
```  
<!-- <font size="1">Columns: RNA_type_Classification </font>      -->
<br/>    

<!-- no Feature peaks with counts > 20 = `r nrow(Peaksdata2_anno[Peaksdata2_anno$Counts_Unique>20&(Peaksdata2_anno$Comb_type_exon_Oppo%in%'no Feature'),])`   -->

<br/>    

```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=8,echo=F,warning=F,eval=T, include=T}
p=Peaksdata2_anno
p$Same_Comb_type_exon=p$Comb_type_exon_Oppo

p=p[p$Same_Comb_type_exon%in%'ncRNA',]

if (nrow(p)>0) {

p$ID_name=paste0(p$ID,"\n",p$Same_gene_name_comb)


colnames(p)[colnames(p)%in%'Same_Comb_type_ncRNA']='RNA_Type'


 gg=ggplot(p,aes(x=RNA_Type,y=Counts_Unique,fill=RNA_Type) )+geom_violin()+theme_classic()+geom_jitter(height = 0, width = 0.1,size=0.5,aes(text=ID_name))+
  scale_y_continuous(trans = "log2")+
  ggtitle('Peak Counts (Unique) for Ro-Clip\nby Feature')+ylab("Counts - Ro-Clip")+xlab("")+
theme(plot.title = element_text(hjust = 0.5,size = 15),axis.title=element_text(size=15),axis.text=element_text(size=12))+theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle = -45))+scale_fill_brewer(palette = "Paired")
 ggplotly(gg,tooltip = c("text", "y"))
}
```  
<br>   
<br>   

#### Fractional Multimap Read counts (Count Unique reads + (MultiMapped read / # of different mappings) )     

Does include rRNA
<br>   

```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T}
p=Peaksdata2_anno
p$Same_Comb_type_exon=p$Comb_type_exon_Oppo


colnames(p)[colnames(p)%in%'Same_Comb_type_exon']='RNA_Type'
p$RNA=''
 gg=ggplot(p,aes(x=RNA_Type,y=Counts_fracMM,fill=RNA_Type) )+geom_violin()+theme_classic()+geom_jitter(height = 0, width = 0.1,size=1)+
  scale_y_continuous(trans = "log2")+
  ggtitle('Peak Counts (Unique + Frac. MM) for Ro-Clip\nby Feature')+ylab("Counts - Ro-Clip")+xlab("")+
theme(plot.title = element_text(hjust = 0.5,size = 25),axis.title=element_text(size=20),axis.text=element_text(size=15))+theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle = -45))

gg
```  
<!-- <font size="1">Columns: RNA_type_Classification </font>      -->
<br/>    
<br/>    

```{r fig.asp=.5,fig.align='center',fig.height=6, fig.width=8,echo=F,warning=F,eval=T}
p=Peaksdata2_anno
p$Same_Comb_type_exon=p$Comb_type_exon_Oppo

p=p[p$Same_Comb_type_exon%in%'ncRNA',]
if (nrow(p)>0) {

colnames(p)[colnames(p)%in%'Same_Comb_type_ncRNA']='RNA_Type'
p$ID_name=paste0(p$ID,"\n",p$Same_gene_name_comb)

 gg=ggplot(p,aes(x=RNA_Type,y=Counts_fracMM,fill=RNA_Type) )+geom_violin()+theme_classic()+geom_jitter(height = 0, width = 0.05,size=.5,aes(text=ID_name))+
  scale_y_continuous(trans = "log2")+
  ggtitle('Peak Counts (Unique + Frac. MM) for Ro-Clip\nby ncRNA Type')+ylab("Counts - Ro-Clip")+xlab("")+
theme(plot.title = element_text(hjust = 0.5,size = 15),axis.title=element_text(size=15),axis.text=element_text(size=12))+theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle = -45))+scale_fill_brewer(palette = "Paired")

 ggplotly(gg,tooltip = c("text", "y"))
 
}
```  
<br>   
<br>   


## Opposite strand CLIP peak attributes   


```{r , fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F}

p=Peaksdata2_anno
p=p[p$Comb_type_exon_Oppo%in%'Antisense Feature',]
if (incd_rRNA==F) {
  p=p[!p$Same_Comb_type_ncRNA%in%'rRNA',]
}

colnames(p)[colnames(p)%in%'Oppo_Comb_type_exon']='RNA_Type'
p$RNA=''

ggplot(p,aes(fill=RNA_Type,x=RNA_Type) )+geom_bar(stat="count")+theme_classic()+
  ggtitle("Opposite Strand Feature - Number of peaks by catagory\n(Determined by unique reads only)")+xlab("")+
theme(plot.title = element_text(hjust = 0.5),axis.title=element_text(size=15),axis.text=element_text(size=10),legend.position = "none",axis.text.x = element_text(angle = -45,vjust = .5))+
 scale_fill_brewer(palette = "Dark2") 

```

```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T,incude=T}

p=Peaksdata2_anno
if (incd_rRNA==F) {
  p=p[!p$Same_Comb_type_ncRNA%in%'rRNA',]
}

p$transition=paste0(p$Same_Comb_type_exon,'->\n',p$Oppo_Comb_type_exon)
p=p[p$Comb_type_exon_Oppo%in%'Antisense Feature',]
colnames(p)[colnames(p)%in%'transition']='RNA_Type'
p$Difference=''

ggplot(p,aes(fill=Same_Comb_type_exon,x=factor(Oppo_Comb_type_exon)) )+geom_bar(stat="count")+theme_classic()+
  ggtitle("Opposite Strand Feature Features: \n Compare Antisense and Snese stand Annotation")+xlab("Antisense strand Annotation")+ylab("Peak Count")+labs(fill = "Sense Stand Annotation")+
theme(plot.title = element_text(hjust = 0.5,size=20),axis.title=element_text(size=15),axis.text=element_text(size=10),legend.position = "right",axis.text.x = element_text(angle = -45,vjust = .5))

```

<br>  

```{r echo=F,warning=F, include=T,eval=T}

p=(Peaksdata2_anno[,grep('Oppo_Repeat_',colnames(Peaksdata2_anno))])
p=unlist(str_split(p,pattern = ","))
p=as.data.frame(count(p[p%in%""==F]))
colnames(p)=c('Repeat_Type','Count')

ggplot(p )+geom_bar(aes(x=Repeat_Type,y=Count),stat="identity")+theme_classic()+#ggtitle("SNP Mutations")+
  xlab("")+ylab("number of peaks in \nrepeat regions")+ggtitle("Opposite Strand")+
theme(plot.title = element_text(hjust = 0.5,size = 15),axis.title=element_text(size=15),axis.text=element_text(size=10),legend.text = element_text( size=10))+theme(axis.text.x = element_text(angle = -45))

```

<br>   
<br> 

###phil repeated section ends

```{r ,fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T,include=F}
###fix variable 
###phil - same as before; if we need to actively call these cols and rename them we should use
#something like select / rename to do it
Peaksdata3_anno=Peaksdata2_anno[,c("ID","IDmerge","strand","chr","start","end","Length",
                                'Counts_Unique','Counts_fracMM',
                                'P_value',paste0('normalized_read_density_in_',samplename),
                                paste0('normalized_read_density_in_',background),'FC',
                                "Same_gene_name_comb","Same_ensembl_gene_id","Same_type_comb",
                                "Same_Repeat_comb",'Same_feature',"Same_Comb_type_exon",
                                "Same_Comb_type_ncRNA","Oppo_gene_name_comb",
                                "Oppo_ensembl_gene_id","Oppo_type_comb","Oppo_Repeat_comb",
                                'Oppo_feature',"Oppo_Comb_type_exon","Oppo_Comb_type_ncRNA",
                                "Comb_type_exon_Oppo")]

colnames(Peaksdata3_anno)=c("ID","IDmerge","strand","chr","start","end" ,"Length",
                          'Counts_Unique', 'Counts_Multimappers_Scaled',
                          'P_value',paste0(samplename,'-normalized_Count'),
                          paste0(background,'-normalized_Count'),'FC',
                          "Same Strand: Host_gene_name","Same Strand: Host_gene_ensembl_id",
                          "Same Strand: RNA_type",'Same Strand: Repeat_Type',
                          "Same Strand: Intron Exon",'Same Strand: RNA_type_Classification',
                          'Same Strand: ncRNA_Classification',"Opposite Strand: Host_gene_name",
                          "Opposite Strand: Host_gene_ensembl_id","Opposite Strand: RNA_type",
                          'Opposite Strand: Repeat_Type',"Opposite Strand: Intron Exon",
                          'Opposite Strand: RNA_type_Classification','Opposite Strand: ncRNA_Classification',
                          'Both Strand: RNA_type_Classification')
###fix variable
Peaksdata4=Peaksdata3_anno

```


# Output Data:  

<br>  

```{r ,fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T,include=F}
#Peaksdata4=Peaksdata4[,colnames(Peaksdata4)[!colnames(Peaksdata4)%in%c('chr','start','end')]]
Peaksdata4=select(Peaksdata4,-c('chr','start','end'))

#output
write.table(Peaksdata4,
            file=output_text, 
            sep = "\t", row.names = FALSE, col.names = T, append = F, quote= FALSE,na = "")

###phil same issue as above
SmplA_Peaks_MAval_out = SmplA_Peaks_MAval[,c("ID","IDmerge","strand","chr","start",
                                             "end" ,"Length",'Counts_Unique',
                                             'Counts_Multimappers_Scaled',
                                             'P_value',
                                             paste0('normalized_read_density_in_',samplename),
                                             paste0('normalized_read_density_in_',background),
                                             'FC', "Same Strand: Host_gene_name",
                                             "Same Strand: Host_gene_ensembl_id","Same Strand: RNA_type",
                                             'Same Strand: Repeat_Type',"Same Strand: Intron Exon",
                                             'Same Strand: RNA_type_Classification',
                                             'Same Strand: ncRNA_Classification',
                                             "Opposite Strand: Host_gene_name","Opposite Strand: Host_gene_ensembl_id",
                                             "Opposite Strand: RNA_type",'Opposite Strand: Repeat_Type',
                                             "Opposite Strand: Intron Exon",'Opposite Strand: RNA_type_Classification',
                                             'Opposite Strand: ncRNA_Classification','Both Strand: RNA_type_Classification')]
#rename, subset
SmplA_Peaks_MAval_out = SmplA_Peaks_MAval_out %>%
    rename(
      !!paste0('normalized_read_density_in_',samplename) :=  paste0(samplename,'-normalized_Count'),
      !!paste0('normalized_read_density_in_',background) :=  paste0(background,'-normalized_Count'),
    ) %>%
  select(-c("chr","start","end"))

#write out
write.table(SmplA_Peaks_MAval_out,
            file=output_text2, 
            sep = "\t", row.names = FALSE, col.names = T, append = F, quote= FALSE,na = "")

```

### Counts Table :    
`r paste0(output_text2)`"   
`r paste0(output_text)`"   

**Columns:**  
"ID" - CLIP peak location  
"strand" - CLIP peak stand  
<!-- "Average MAPQ" - Average Mapping quality of reads aligned to peak region.   -->
"Counts_Unique" - Number of Unique reads aligned to CLIP peak  
<!-- "Counts_All" - Count of Multimapped reads for each peak     -->
<!-- "Counts_All" - Count of Unique + Multimapped read for each peak (Each occurrence of Multimapped read is counted separately)   -->
"Counts_Multimappers_Scaled"  - number of reads with Unique reads =1 and Multimapped =1/#MM locations  
<!-- "Counts_Multimappers_Best Mapping" - Number of Unique reads + best mapQ Multimapped Read aligned to CLIP peak   -->
<!-- "Multimaped Reads %" - Percent of Multimapped reads that align to CLIP peak   -->
<!-- "Pvalue" - Sample specific peak P-Value compared to Control - MAnorm.   -->
"`r samplename`-normalized_Count"- Normalized Peak count generated by MAnorm.  
"`r background`-normalized_Count"- Control Sample normalized Peak count generated by MAnorm.  
"FC" - Log2 Fold Change (`r samplename`,-`r background`) using MAnorm normalized counts.  
  
"Same Strand: Host_gene_name" - Name of Host Gene on same strand as CLIP reads  
"Same Strand: Host_gene_ensembl_id" - Ensemble Name of Host Gene on same strand as CLIP reads  
"Same Strand: RNA_type" - Transcript type of CLIP peak host gene on same strand as CLIP reads  
"Same Strand: Repeat_Type" - Type of repeat in Clip peak (determined by repeat masker) on same strand as CLIP reads  
"Same Strand: Intron Exon" - Gene feature of CLIP peak location in protein coding Gene on same strand as CLIP reads  
"Same Strand: RNA_type_Classification" - Summary of Peak Classification   
"Same Strand: ncRNA_Classification" - Summary of ncRNA Peak assignments  

"Opposite Strand: Host_gene_name" - Name of Host Gene on opposite strand as CLIP reads  
"Opposite Strand: Host_gene_ensembl_id" - Ensemble Name of Host Gene on opposite strand as CLIP reads  
"Opposite Strand: RNA_type" - Transcript type of CLIP peak host gene on opposite strand as CLIP reads  
"Opposite Strand: Repeat_Type" - Type of repeat in Clip peak (determined by repeat masker) on opposite strand as CLIP reads  
"Opposite Strand: Intron Exon" - Gene feature of CLIP peak location in protein coding Gene on opposite strand as CLIP reads  
"Opposite Strand: RNA_type_Classification" - Summary of Peak Classification  
"Opposite Strand: ncRNA_Classification" - Summary of ncRNA Peak assignments  

"Both Strand: RNA_type_Classification" - Summary of Peak Classification including both strand annotations  
      

<!-- "MM_number" - # of reads in CLIP peak that aligns to other locations Genome   -->
<!-- "MM_Alt_Peaktype" - # of Multimapped reads with Alternative Peak Classifications different from original Peak Classification   -->
<!-- "MM_Same_Peaktype" - # of Multimapped reads with the same Peak Classification as the original Peak Classification   -->
<!-- "MM_type" - Peak Classifications for alternative mappings of multimapped reads   -->
<!-- "MM_Prec_Alt_Peaktype" - % of Multimapped reads with Alternative Peak Classifications different from original Peak Classification   -->
<!-- "MM_Alt_Alignments" - Location alternate alignments for Multimapped reads   -->

<!-- "iCount_peak" - 1= peak is identified by iCount -->
<br>

<!-- ## Alignment Tracks   -->

<!-- `r toString(samplename)`_Clip_iCountcutadpt_all.unique.NH.mm.ddup.bam - MultiMapped + Uniquely aligned CLIP Reads   -->
<!-- `r toString(samplename)`_Clip_iCountcutadpt_all.unique.NH.mm.ddup.unique.bam - Uniquely aligned CLIP Reads   -->
<!-- `r toString(samplename)`_Clip_iCountcutadpt_all.unique.NH.mm.ddup.multimapped.bam - MultiMapped aligned CLIP Reads   -->

<br>
<br>

<!-- ## Annotation Tracks     -->

<!-- sftp://Biowulf.nih.gov/data/RBL_NCI/datashare/wolin/mESC_Clip/annotation_files/ -->


<!-- <br>   -->
<!-- <br>   -->















