#swarm -f hg38_scriptloop_MakeTransctome_SplicedTransc_46R.sh --verbose 1 -g 50 -t 32 -b 1 --time 10-00:00:00 --job-name 46R_SplicedTransc_MakeTransctome_mm10_rRNA --module samtools,novocraft

java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270337.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270708.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270728.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270364.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ458384.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270448.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KN196475.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270835.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ458385.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KZ559113.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL949749.2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.chr17.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL000221.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383575.2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL000219.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270749.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143355.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KZ208908.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270722.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ090025.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270312.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270846.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ090017.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL000255.2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383565.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KV575250.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KN538364.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ090020.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143378.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.chr9.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270867.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383579.2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.chr13.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270721.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270856.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270419.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143384.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143369.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270802.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383522.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270857.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383542.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL000254.2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270465.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ031390.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KN538370.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270788.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383571.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ090024.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270770.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143352.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.JH159148.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270322.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270438.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL000194.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270582.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270780.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270384.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143360.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270748.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KZ559100.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270906.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270778.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ983256.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383577.2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270920.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270777.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270385.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ031385.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL000225.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383569.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270917.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270779.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.chr19.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270509.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL877876.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383556.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270832.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270769.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270335.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383578.2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KN196476.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270302.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270807.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270913.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270588.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL000226.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL000218.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270731.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383539.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270548.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383534.2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270842.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383581.2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.chr8.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270859.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270414.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270899.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270394.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383566.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KZ208905.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270718.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143370.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143341.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ090022.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ759761.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270907.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KV880765.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KV575254.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383573.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270889.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270885.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.chr12.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383546.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270310.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270783.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270866.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270379.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270921.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KZ208921.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383582.2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ983258.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KN538363.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KN196483.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KV575243.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270539.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383527.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KV575252.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270918.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270887.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270830.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270751.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270418.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270362.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270831.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270912.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143350.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270420.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270522.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270904.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270744.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KV766196.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KV575244.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270790.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270879.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143357.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270758.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270902.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270392.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270876.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ090023.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270706.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KV880764.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.chr7.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270726.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270736.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143373.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL000214.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270746.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ031386.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270765.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL339449.2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270511.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KZ559106.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270865.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270894.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270771.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270935.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383564.2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270737.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270754.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143381.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL000250.2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270709.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KN196472.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383576.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.chr14.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270512.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL000213.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143377.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270798.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270727.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270872.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270927.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ090028.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270820.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270817.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143353.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL949750.2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270317.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270855.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270714.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ090018.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KV766197.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KN538360.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270871.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383572.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ031388.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KZ559112.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270374.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL000251.2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270382.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270435.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270466.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270764.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270847.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383563.3.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270716.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KZ559111.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KZ208910.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KV880768.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270838.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270741.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270815.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KV766198.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270809.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL000205.2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270583.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KV575247.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270837.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143372.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KV766195.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KN538367.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL949747.2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.JH159146.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270581.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.chr6.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270850.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270338.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KN196478.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270711.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KV575253.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KZ559109.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ090013.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KV575255.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270390.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270729.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270873.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270928.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143346.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270393.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270784.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143379.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383551.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270762.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL000253.2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270733.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ090016.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270378.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143364.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KZ208923.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270544.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ759762.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270936.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143367.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ090027.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270795.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KV575256.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KV575245.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270827.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143383.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270916.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383521.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270386.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270853.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270852.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL000224.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KN538371.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270882.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KZ208922.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270908.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270710.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270707.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143366.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270819.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270905.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270789.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270723.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270372.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KN196473.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383530.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270763.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270468.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270442.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143365.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270753.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270768.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270862.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270373.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KN196485.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270775.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143374.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270521.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383550.2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ090021.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270761.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.chr5.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383540.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KZ559114.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ090026.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270808.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270915.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270814.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270745.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KB663609.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270742.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270812.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270870.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270589.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL000220.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KV766199.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ090015.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270914.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270864.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KZ208920.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270516.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270929.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383549.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270823.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270888.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KN538365.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143356.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383531.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270760.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270881.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KN196487.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270825.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270396.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143376.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270839.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270877.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270732.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270508.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270821.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270518.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383518.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270932.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL000008.2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383526.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.chr22.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270412.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.chr15.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL949748.2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383545.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383533.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ458386.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270806.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270886.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270840.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270528.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KN196481.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KV880766.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270919.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270818.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383583.2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270843.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383580.2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KN196486.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KZ559108.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270329.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270799.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL000216.2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270923.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270743.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383552.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.chr21.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ090014.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270786.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.chr4.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ458382.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270591.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270804.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143354.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143359.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KV575248.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270895.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ983257.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270730.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270776.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.JH159137.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270822.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL949753.2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270580.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143349.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270875.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270340.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270593.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.chr20.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270739.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270752.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270937.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143371.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KZ559102.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143375.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270863.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270848.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ031383.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270834.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270774.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270891.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270530.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270391.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL949751.2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270712.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270910.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ031389.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270587.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270395.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KV880767.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270515.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KZ208915.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ031384.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270787.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270909.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ458388.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270874.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ090019.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL582966.2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270336.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KV575259.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KV766192.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KZ559116.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270756.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270717.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270772.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383519.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270938.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270794.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270363.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383570.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KZ559103.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KZ559105.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270933.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KV575249.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KZ208918.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ759760.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143342.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270510.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270330.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.chr3.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KN196480.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270311.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270429.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270810.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270833.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270467.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL949752.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270316.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270538.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KZ208913.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.chrY.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383547.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270371.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270896.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.chr10.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270849.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KZ208919.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270303.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270773.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KZ208906.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143358.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143362.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270719.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KV575246.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270734.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270844.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270930.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270755.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.JH159136.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270845.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270883.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KN538366.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KZ208912.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270529.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270785.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270519.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143351.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143385.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KB021644.2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270851.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270423.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383532.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270334.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KZ208907.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383528.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KN538373.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KN538372.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270801.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KN196484.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KZ208924.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270781.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270422.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL000258.2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270738.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270893.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383553.2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270305.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270304.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270892.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270579.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.chr16.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270898.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KZ208916.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270841.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270713.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270791.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270797.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.chr2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383555.2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270880.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270824.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383567.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270724.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KZ559101.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270381.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KZ208911.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KV575251.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.chrX.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270925.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KZ208909.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ031387.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270747.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270720.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143343.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270333.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270750.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270858.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KZ208904.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270715.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KV880763.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL000195.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270931.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL000256.2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143347.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270934.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270869.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270897.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.JH159147.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270813.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143382.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383554.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383557.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL949742.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270792.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL000257.2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270811.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143344.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270767.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270922.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270924.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270793.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KN538361.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270836.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.chr18.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270826.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KZ559107.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270890.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270884.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270375.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270800.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270816.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143361.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ983255.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270517.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.chr1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383574.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383568.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KV575258.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KZ559115.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KZ559110.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270376.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270901.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270759.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383541.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270868.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KZ559104.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270507.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KN538369.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270417.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL877875.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270803.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270757.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KZ208914.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KV766193.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270900.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ458383.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270383.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270796.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KN538368.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270782.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270740.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270315.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143345.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KV575260.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270366.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL000209.2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270878.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270389.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270766.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143363.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143348.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.JH636055.2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270411.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KN196474.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270424.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ458387.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL949746.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270805.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.chr11.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270735.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270425.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KN538362.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KN196482.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KN196479.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270320.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.chrM.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KN196477.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270860.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KQ759759.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL000252.2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270590.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143368.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270911.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270387.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270861.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270854.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270829.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL000009.2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL000208.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270388.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270903.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.GL383520.2.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KV575257.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270584.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KZ208917.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.ML143380.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270926.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KI270725.1.phred2 \
-r 46 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation.gtf.SplicedTransc.KV766194.1.phred2 \
-r 46 -s 




# mv /data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/splitphred_SplicedTransc/*.gz /data/data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/MakeTranscriptome_SplicedTransc_46R 
# gunzip /data/data/RBL_NCI/Phil/Reference/Index/novoalign/hg38/hg38_gencode_v32/MakeTranscriptome_genome/MakeTranscriptome_SplicedTransc_46R/*.gz 
