#swarm -f mm10_rRNA_scriptloop_MakeTransctome_SplicedTransc_71R.sh --verbose 1 -g 50 -t 32 -b 1 --time 10-00:00:00 --job-name 71R_SplicedTransc_MakeTransctome_mm10_rRNA --module samtools,novocraft

java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KZ289093.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KV575242.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.JH792832.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.GL456359.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.chr17.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.chr9.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KZ289075.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.chr13.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KV575241.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.JH792834.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.JH792826.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.GL456213.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.GL456210.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KQ030491.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KV575240.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.GL456366.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KZ289066.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KQ030485.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.chr19.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KZ289091.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KZ289065.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KK082442.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.GL456372.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.chr8.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KV575236.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KQ030494.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.GL456221.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KZ289095.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.GL456382.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.JH584293.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.JH792829.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.chr12.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KZ289094.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.BK000964.3.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KQ030484.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KB469740.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.GL456394.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.GL456379.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.JH792828.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.GL456354.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.chr7.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.JH584297.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KZ289072.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.chr14.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.JH584300.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KK082441.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.GL456383.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KZ289068.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.JH792827.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KZ289070.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.GL456378.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KZ289092.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KZ289087.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.chr6.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KB469741.2.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KZ289085.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KZ289069.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.GL456212.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KV575233.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KZ289078.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KB469738.3.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KZ289086.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.JH792830.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KQ030492.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.GL456393.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.chr5.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KZ289082.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KZ289064.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.GL456381.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.JH584299.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KZ289071.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.GL456211.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.JH584302.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KV575239.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KZ289079.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.JH584298.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.chr15.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.GL456389.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.JH584292.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KB469739.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.chr4.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.GL456392.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.GL456396.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KQ030487.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.GL456239.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KV575238.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.GL456216.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KV575232.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KZ289074.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KQ030490.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.GL456233.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.JH584304.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KQ030495.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KQ030493.2.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.chr3.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.GL456368.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KZ289073.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.chrY.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KZ289088.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.chr10.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KZ289090.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KB469742.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.JH792831.2.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KZ289083.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KZ289084.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.chr16.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KZ289080.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.chr2.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.chrX.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.GL456350.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.JH584295.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KZ289076.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.GL456385.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.GL456367.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KZ289077.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.JH792833.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.GL456219.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KQ030486.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KZ289089.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.JH584294.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.chr18.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KQ030489.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KZ289081.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.chr1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KZ289067.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.GL456390.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KV575234.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.GL456370.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KQ030488.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KQ030496.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.JH584301.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.GL456360.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.JH584296.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.JH584303.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.chr11.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.chrM.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.GL456387.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KQ030497.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KV575235.1.phred2 \
-r 71 -s 


java -Xmx100G -jar /data/RBL_NCI/Phil/Tools/USeq_8.9.6/Apps/MakeTranscriptome \
-f /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/genome/ \
-u /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/gencode.v32.chr_patch_hapl_scaff.annotation_BK000964.3.bed.SplicedTransc.KV575237.1.phred2 \
-r 71 -s 




# mv /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/*Rad71Num100kMin10Splices.fasta.gz /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/MakeTranscriptome_SplicedTransc_71R/Chr_split/
# mv /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/splitphred_SplicedTransc/*Rad71Num100kMin10Transcripts.fasta.gz /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/MakeTranscriptome_SplicedTransc_71R/Chr_split/
# gunzip /data/RBL_NCI/Phil/Reference/Index/novoalign/mm10_rRNA/MakeTranscriptome_genome/MakeTranscriptome_SplicedTransc_71R/Chr_split/*.gz
